# 第5章 単層ネットワーク: 分類 (Single-layer Networks: Classification)
## 章末演習問題 (Exercises 5.1 〜 5.24)

本ノートブックは、教科書『深層学習：基礎と概念 (Deep Learning: Foundations and Concepts)』(Christopher M. Bishop & Hugh Bishop 著, 2024年刊) の **第5章「単層ネットワーク: 分類」章末演習問題 全24問 (Exercises 5.1 〜 5.24)** の完全解答・数式厳密導出・自己検証コードです。

---

### 演習問題 目次
- **Part 1: 5.1節 線形判別モデル (Linear Discriminant Models) [Ex 5.1 - 5.4]**
  - **Exercise 5.1 (?)**: 1-of-$K$ 二値符号化における条件付き期待値 $\mathbb{E}[\mathbf{t} \mid \mathbf{x}] = p(\mathcal{C}_k \mid \mathbf{x})$ の証明
  - **Exercise 5.2 (? ?)**: 凸包 (Convex Hull) の交差と線形分離可能性の同値性証明（線形計画法）
  - **Exercise 5.3 (? ?)**: 二乗和誤差最小化における線形制約 $\mathbf{a}^T \mathbf{t}_n + b = 0$ の予測値への保存（バイアス項 $\phi_0(\mathbf{x})=1$ の役割）
  - **Exercise 5.4 (? ?)**: 複数の同時線形制約 $\mathbf{A}^T \mathbf{t}_n + \mathbf{b} = \mathbf{0}$ の最小二乗予測値への保存
- **Part 2: 5.2節 決定理論 (Decision Theory for Classification) [Ex 5.5 - 5.10]**
  - **Exercise 5.5 (?)**: 適合率 (Precision) と再現率 (Recall) からの $F_\beta$ スコア公式の厳密導出
  - **Exercise 5.6 (? ?)**: バタチャリア境界 (Bhattacharyya Bound): 誤分類確率の上界 $p(\text{mistake}) \le \int \{p(\mathbf{x}, \mathcal{C}_1)p(\mathbf{x}, \mathcal{C}_2)\}^{1/2} d\mathbf{x}$
  - **Exercise 5.7 (?)**: 0-1 損失行列 $L_{kj} = 1 - I_{kj}$ による期待リスク最小化が事後確率最大化に帰着することの証明
  - **Exercise 5.8 (?)**: 一般損失行列 $L_{kj}$ と事前確率 $\pi_k$ の下での期待損失最小化決定基準
  - **Exercise 5.9 (?)**: サンプル事後確率の標本平均の事前確率 $p(\mathcal{C}_k)$ への大数の法則による漸近収束
  - **Exercise 5.10 (? ?)**: 棄却オプション (Reject Option) 付き決定基準と閾値関係式 $\theta = 1 - \lambda$ の導出
- **Part 3: 5.3節 生成モデル (Generative Classifiers) [Ex 5.11 - 5.17]**
  - **Exercise 5.11 (?)**: ロジスティック・シグモイド関数の対称性 $\sigma(-a) = 1 - \sigma(a)$ と逆関数ロジット $\sigma^{-1}(y) = \ln(y/(1-y))$
  - **Exercise 5.12 (?)**: 共通共分散ガウス生成モデルにおける事後確率ロジスティック形式とパラメータ $\mathbf{w}, w_0$ の解析解
  - **Exercise 5.13 (?)**: $K$ クラス生成モデルの事前確率 $\pi_k = N_k / N$ のラグランジュ未定乗数法による最尤推定
  - **Exercise 5.14 (? ?)**: 共通共分散多変量ガウス生成モデルのクラス平均 $\boldsymbol{\mu}_k$ および共通共分散行列 $\boldsymbol{\Sigma}$ の最尤推定解
  - **Exercise 5.15 (? ?)**: 離散2値特徴ベルヌーイナイーブベイズの最尤推定量 $\mu_{ki} = \frac{1}{N_k}\sum t_{nk} x_{ni}$
  - **Exercise 5.16 (? ?)**: $L$ 状態離散特徴をもつマルチステートナイーブベイズの事前活性化 $a_k$ の線形性証明
  - **Exercise 5.17 (? ?)**: マルチステートナイーブベイズパラメータ $\mu_{kml} = N_{kml} / N_k$ のラグランジュ乗数法による最尤推定
- **Part 4: 5.4節 識別モデル (Discriminative Classifiers) [Ex 5.18 - 5.24]**
  - **Exercise 5.18 (?)**: ロジスティック・シグモイド関数の導関数公式 $\frac{d\sigma}{da} = \sigma(1 - \sigma)$ の検証
  - **Exercise 5.19 (?)**: ロジスティック回帰の交差エントロピー誤差関数の勾配 $\nabla E = \sum_n (y_n - t_n)\boldsymbol{\phi}_n$ の導出
  - **Exercise 5.20 (?)**: 線形分離可能データにおける非正則化ロジスティック回帰の重みノルム発散 $\|\mathbf{w}\| \to \infty$ の証明
  - **Exercise 5.21 (?)**: ソフトマックス活性化関数のヤコビ行列 $\frac{\partial y_k}{\partial a_j} = y_k(I_{kj} - y_j)$ の厳密導出
  - **Exercise 5.22 (?)**: 多クラス交差エントロピー誤差関数の勾配 $\nabla_{\mathbf{w}_j} E = \sum_n (y_{nj} - t_{nj})\boldsymbol{\phi}_n$ の導出
  - **Exercise 5.23 (?)**: プロビット関数 $\Phi(a)$ と誤差関数 $\operatorname{erf}(x)$ の恒等式 $\Phi(a) = \frac{1}{2}[1 + \operatorname{erf}(a/\sqrt{2})]$ の証明
  - **Exercise 5.24 (? ?)**: ロジスティック・シグモイド $\sigma(a)$ とスケールドプロビット $\Phi(\lambda a)$ の原点傾き一致条件 $\lambda^2 = \pi / 8$ の導出


In [1]:
# 環境セットアップと共通ライブラリのインポート
import os
import sys
import numpy as np
import scipy.integrate as integrate
import scipy.special as special
from scipy.optimize import linprog
from scipy.stats import norm
import matplotlib.pyplot as plt

# プロジェクトルートの設定
current_dir = os.getcwd()
if os.path.basename(current_dir) == '5':
    repo_root = os.path.abspath(os.path.join(current_dir, '..'))
else:
    repo_root = os.path.abspath(current_dir)

if repo_root not in sys.path:
    sys.path.insert(0, repo_root)

from common.plot_utils import setup_style
from common.discriminative_classifiers import (
    sigmoid, sigmoid_deriv, logit, softmax, softmax_jacobian,
    probit, probit_deriv, erf_func, LogisticRegression, SoftmaxRegression, ProbitRegression
)
from common.exercises_ch5 import (
    exercise_5_1_conditional_expectation,
    exercise_5_2_check_separability,
    exercise_5_3_verify_linear_constraint,
    exercise_5_4_verify_multiple_constraints,
    exercise_5_5_compute_f_score,
    exercise_5_6_compute_bhattacharyya_bound,
    exercise_5_7_zero_one_loss_decision,
    exercise_5_8_expected_loss_decision,
    exercise_5_9_average_posterior,
    exercise_5_10_reject_threshold,
    exercise_5_11_verify_sigmoid_properties,
    exercise_5_12_compute_gaussian_gda_params,
    exercise_5_13_mle_priors,
    exercise_5_14_mle_gaussian_shared_cov,
    exercise_5_15_mle_binary_naive_bayes,
    exercise_5_16_multistate_naive_bayes_ak,
    exercise_5_17_mle_multistate_naive_bayes,
    exercise_5_18_verify_sigmoid_derivative,
    exercise_5_19_verify_logistic_gradient,
    exercise_5_20_separable_growth_check,
    exercise_5_21_verify_softmax_jacobian,
    exercise_5_22_verify_softmax_gradient,
    exercise_5_23_verify_probit_erf,
    exercise_5_24_probit_sigmoid_matching_scale
)

setup_style()
print("第5章 演習問題: セットアップが正常に完了しました。")


第5章 演習問題: セットアップが正常に完了しました。


## Part 1: 5.1節 線形判別モデル (Linear Discriminant Models)


---
### Exercise 5.1: 1-of-K 表現における条件付き期待値
**Problem 5.1 (?)**:
> Consider a classification problem with $K$ classes and a target vector $\mathbf{t}$ that uses a 1-of-$K$ binary coding scheme. Show that the conditional expectation $\mathbb{E}[\mathbf{t} \mid \mathbf{x}]$ is given by the posterior probability $p(\mathcal{C}_k \mid \mathbf{x})$.

#### 数式導出と論理ステップ:
1. **1-of-$K$ 二値符号化スキーマの定義**:
   $K$ クラス分類において、目標ベクトル $\mathbf{t}$ は入力 $\mathbf{x}$ がクラス $\mathcal{C}_k$ に属するとき、第 $k$ 成分のみが 1 で他が 0 である標準基底ベクトル $\mathbf{e}_k = (0, \dots, 1, \dots, 0)^T \in \mathbb{R}^K$ をとります：
   $$
   \mathbf{t} = \mathbf{e}_k \iff \mathbf{x} \in \mathcal{C}_k
   $$
   入力 $\mathbf{x}$ が与えられたときの目標値 $\mathbf{t}$ の条件付き確率分布は離散分布であり：
   $$
   p(\mathbf{t} = \mathbf{e}_k \mid \mathbf{x}) = p(\mathcal{C}_k \mid \mathbf{x})
   $$

2. **条件付き期待値の計算**:
   離散確率変数の期待値の定義より：
   $$
   \mathbb{E}[\mathbf{t} \mid \mathbf{x}] = \sum_{k=1}^K \mathbf{e}_k \, p(\mathbf{t} = \mathbf{e}_k \mid \mathbf{x}) = \sum_{k=1}^K \mathbf{e}_k \, p(\mathcal{C}_k \mid \mathbf{x})
   $$
   各成分 $j \in \{1, \dots, K\}$ について考えると：
   $$
   \mathbb{E}[t_j \mid \mathbf{x}] = \sum_{k=1}^K (\mathbf{e}_k)_j \, p(\mathcal{C}_k \mid \mathbf{x}) = \sum_{k=1}^K I_{jk} \, p(\mathcal{C}_k \mid \mathbf{x}) = p(\mathcal{C}_j \mid \mathbf{x})
   $$
   したがって、ベクトル全体として：
   $$
   \mathbb{E}[\mathbf{t} \mid \mathbf{x}] = \begin{pmatrix} p(\mathcal{C}_1 \mid \mathbf{x}) \\ \vdots \\ p(\mathcal{C}_K \mid \mathbf{x}) \end{pmatrix}
   $$
   が成り立ちます。これにより、二乗和誤差を最小化する回帰モデル $y_k(\mathbf{x})$ の最適予測値がベイズ事後確率 $p(\mathcal{C}_k \mid \mathbf{x})$ を近似することが示されました。 $\blacksquare$


In [2]:
# Exercise 5.1 数値検証
p_true = np.array([0.15, 0.55, 0.30])
E_t = exercise_5_1_conditional_expectation(p_true)

print("真の事後確率ベクトル p(C_k|x):", p_true)
print("計算された条件付き期待値 E[t|x]:", E_t)

assert np.allclose(E_t, p_true), "条件付き期待値は事後確率と完全に一致する必要があります。"
print("Exercise 5.1: 検証成功！")


真の事後確率ベクトル p(C_k|x): [0.15 0.55 0.3 ]
計算された条件付き期待値 E[t|x]: [0.15 0.55 0.3 ]
Exercise 5.1: 検証成功！


---
### Exercise 5.2: 凸包の交差と線形分離可能性の同値性
**Problem 5.2 (? ?)**:
> Given a set of data points $\{\mathbf{x}_n\}$, we can define the convex hull to be the set of all points $\mathbf{x}$ given by $\mathbf{x} = \sum_n \alpha_n \mathbf{x}_n$ where $\alpha_n \ge 0$ and $\sum_n \alpha_n = 1$. Consider a second set of points $\{\mathbf{y}_m\}$ together with their corresponding convex hull. Show that if their convex hulls intersect, the two sets of points cannot be linearly separable, and conversely that if they are linearly separable, their convex hulls do not intersect.

#### 数式導出と論理ステップ:
1. **凸包 (Convex Hull) の定義**:
   点集合 $X = \{\mathbf{x}_n\}_{n=1}^{N_x}$ および $Y = \{\mathbf{y}_m\}_{m=1}^{N_y}$ の凸包は次で定義されます：
   $$
   \operatorname{conv}(X) = \left\{ \sum_{n=1}^{N_x} \alpha_n \mathbf{x}_n \;\middle|\; \alpha_n \ge 0, \; \sum_{n=1}^{N_x} \alpha_n = 1 \right\}, \quad \operatorname{conv}(Y) = \left\{ \sum_{m=1}^{N_y} \beta_m \mathbf{y}_m \;\middle|\; \beta_m \ge 0, \; \sum_{m=1}^{N_y} \beta_m = 1 \right\}
   $$

2. **線形分離可能性の定義**:
   $X$ と $Y$ が線形分離可能であるとは、法線ベクトル $\mathbf{w} \in \mathbb{R}^D$ とバイアス $w_0 \in \mathbb{R}$ が存在して：
   $$
   \forall n, \; \mathbf{w}^T \mathbf{x}_n + w_0 > 0 \quad \text{かつ} \quad \forall m, \; \mathbf{w}^T \mathbf{y}_m + w_0 < 0
   $$
   が満たされることを言います。

3. **順方向の背理法証明 (凸包が交差する $\implies$ 線形分離不可能)**:
   凸包が交差すると仮定します。すなわち、ある点 $\mathbf{z} \in \operatorname{conv}(X) \cap \operatorname{conv}(Y)$ が存在します：
   $$
   \mathbf{z} = \sum_{n=1}^{N_x} \alpha_n \mathbf{x}_n = \sum_{m=1}^{N_y} \beta_m \mathbf{y}_m
   $$
   背理法のため、$X$ と $Y$ が分離超平面 $(\mathbf{w}, w_0)$ で線形分離可能であると仮定します。
   各 $\mathbf{x}_n$ の不等式に $\alpha_n \ge 0$ を掛けて総和をとると（少なくとも1つの $\alpha_n > 0$）：
   $$
   \sum_{n=1}^{N_x} \alpha_n (\mathbf{w}^T \mathbf{x}_n + w_0) = \mathbf{w}^T \left( \sum_{n=1}^{N_x} \alpha_n \mathbf{x}_n \right) + w_0 \left(\sum_{n=1}^{N_x} \alpha_n\right) = \mathbf{w}^T \mathbf{z} + w_0 > 0
   $$
   同様に各 $\mathbf{y}_m$ の不等式に $\beta_m \ge 0$ を掛けて総和をとると：
   $$
   \sum_{m=1}^{N_y} \beta_m (\mathbf{w}^T \mathbf{y}_m + w_0) = \mathbf{w}^T \left( \sum_{m=1}^{N_y} \beta_m \mathbf{y}_m \right) + w_0 \left(\sum_{m=1}^{N_y} \beta_m\right) = \mathbf{w}^T \mathbf{z} + w_0 < 0
   $$
   これは $0 < \mathbf{w}^T \mathbf{z} + w_0 < 0$ という明白な矛盾を導きます。したがって、凸包が交差する場合、データは線形分離不可能です。

4. **逆方向の証明 (線形分離可能 $\implies$ 凸包が交差しない / 凸包が非交差 $\implies$ 線形分離可能)**:
   対偶「凸包が交差しない $\implies$ 線形分離可能」を示します。
   有限集合の凸包 $\operatorname{conv}(X)$ および $\operatorname{conv}(Y)$ は、$\mathbb{R}^D$ の有界閉集合（コンパクト集合）かつ凸集合です。
   $\operatorname{conv}(X) \cap \operatorname{conv}(Y) = \emptyset$ であるとき、**強分離超平面定理 (Strong Separating Hyperplane Theorem)** より、超平面 $(\mathbf{w}, w_0)$ が存在して：
   $$
   \forall \mathbf{x} \in \operatorname{conv}(X), \; \mathbf{w}^T \mathbf{x} + w_0 > 0 \quad \text{かつ} \quad \forall \mathbf{y} \in \operatorname{conv}(Y), \; \mathbf{w}^T \mathbf{y} + w_0 < 0
   $$
   特に $X \subset \operatorname{conv}(X)$ および $Y \subset \operatorname{conv}(Y)$ より、元の点集合も厳密に線形分離されます。 $\blacksquare$


In [3]:
# Exercise 5.2 数値検証（線形計画法による分離可能性判定）
# ケースA: 互いに交差しない点群（線形分離可能）
X_sep = np.array([[1.0, 1.0], [2.0, 1.5], [1.5, 2.5]])
Y_sep = np.array([[-1.0, -1.0], [-2.0, -0.5], [-1.2, -2.0]])
sep_result_A = exercise_5_2_check_separability(X_sep, Y_sep)

# ケースB: 一方が他方の凸包の内部に存在する点群（交差・分離不可能）
X_cross = np.array([[0.0, 0.0]])
Y_cross = np.array([[1.0, 1.0], [-1.0, 1.0], [1.0, -1.0], [-1.0, -1.0]])
sep_result_B = exercise_5_2_check_separability(X_cross, Y_cross)

print("非交差データセットの分離可能性判定:", sep_result_A)
print("交差データセットの分離可能性判定:", sep_result_B)

assert sep_result_A is True, "非交差凸包は線形分離可能でなければなりません。"
assert sep_result_B is False, "交差凸包は線形分離不可能でなければなりません。"
print("Exercise 5.2: 検証成功！")


非交差データセットの分離可能性判定: True
交差データセットの分離可能性判定: False
Exercise 5.2: 検証成功！


---
### Exercise 5.3: 最小二乗予測値における線形制約の保存
**Problem 5.3 (? ?)**:
> Consider the minimization of a sum-of-squares error function (5.14), and suppose that all the target vectors in the training set satisfy a linear constraint $\mathbf{a}^T \mathbf{t}_n + b = 0$. Show that as a consequence of this constraint, the elements of the model prediction $\mathbf{y}(\mathbf{x})$ given by the least-squares solution (5.16) also satisfy this constraint, so that $\mathbf{a}^T \mathbf{y}(\mathbf{x}) + b = 0$. Assume that one of the basis functions $\phi_0(\mathbf{x}) = 1$ so that the corresponding parameter $\mathbf{w}_0$ plays the role of a bias.

#### 数式導出と論理ステップ:
1. **行列形式における線形制約の定式化**:
   訓練目標行列 $\mathbf{T} \in \mathbb{R}^{N \times K}$ の各行 $\mathbf{t}_n^T$ が制約 $\mathbf{a}^T \mathbf{t}_n + b = 0$ を満たすとき、すべての行ベクトルについて転置をとると $\mathbf{t}_n^T \mathbf{a} + b = 0$ です。
   全訓練データ $N$ 個をまとめて行列表記すると：
   $$
   \mathbf{T} \mathbf{a} + b \mathbf{1}_N = \mathbf{0}_N \iff \mathbf{T} \mathbf{a} = -b \mathbf{1}_N
   $$
   ここで $\mathbf{1}_N = (1, 1, \dots, 1)^T \in \mathbb{R}^N$ です。

2. **計画行列におけるバイアス列の性質**:
   基底関数に定数項 $\phi_0(\mathbf{x}) = 1$ が含まれているため、計画行列 $\boldsymbol{\Phi} \in \mathbb{R}^{N \times M}$ の第 1 列（インデックス 0）はすべて 1 です：
   $$
   \boldsymbol{\Phi} \mathbf{e}_1 = \mathbf{1}_N, \quad \text{ただし } \mathbf{e}_1 = (1, 0, \dots, 0)^T \in \mathbb{R}^M
   $$

3. **最小二乗解 $\mathbf{W}$ への制約ベクトルの作用**:
   多変量最小二乗解 (式 5.16) は次式で与えられます：
   $$
   \mathbf{W} = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T \mathbf{T}
   $$
   これに右からベクトル $\mathbf{a} \in \mathbb{R}^K$ を掛け合わせます：
   $$
   \mathbf{W} \mathbf{a} = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T (\mathbf{T} \mathbf{a}) = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T (-b \mathbf{1}_N)
   $$
   ここで $\mathbf{1}_N = \boldsymbol{\Phi} \mathbf{e}_1$ を代入すると：
   $$
   \mathbf{W} \mathbf{a} = -b (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} (\boldsymbol{\Phi}^T \boldsymbol{\Phi}) \mathbf{e}_1 = -b \mathbf{e}_1
   $$
   という極めて簡潔なベクトルが得られます。

4. **任意のテスト入力 $\mathbf{x}$ に対する予測値の検証**:
   任意の入力点 $\mathbf{x}$ に対するモデルの予測値は $\mathbf{y}(\mathbf{x}) = \mathbf{W}^T \boldsymbol{\phi}(\mathbf{x})$ です。
   制約式の左辺を計算します：
   $$
   \mathbf{a}^T \mathbf{y}(\mathbf{x}) = \mathbf{y}(\mathbf{x})^T \mathbf{a} = \boldsymbol{\phi}(\mathbf{x})^T \mathbf{W} \mathbf{a} = \boldsymbol{\phi}(\mathbf{x})^T (-b \mathbf{e}_1) = -b \phi_0(\mathbf{x})
   $$
   $\phi_0(\mathbf{x}) = 1$ であるため：
   $$
   \mathbf{a}^T \mathbf{y}(\mathbf{x}) = -b \cdot 1 = -b \implies \mathbf{a}^T \mathbf{y}(\mathbf{x}) + b = 0
   $$
   となり、訓練データに課された任意の線形制約が、最小二乗モデルのあらゆる未知の予測点 $\mathbf{x}$ においても数学的恒等式として完全に保存されることが証明されました。 $\blacksquare$


In [4]:
# Exercise 5.3 数値検証
np.random.seed(42)
N, D, K = 50, 3, 4
X_train = np.random.randn(N, D)
a_vec = np.array([2.0, -1.0, 3.0, 0.5])
b_const = -4.5

# 制約 a^T t_n + b = 0 を満たす目標値 T を生成
T_train = np.random.randn(N, K)
T_train[:, -1] = (-b_const - T_train[:, :-1] @ a_vec[:-1]) / a_vec[-1]

X_test = np.random.randn(20, D)
train_viol, test_viol = exercise_5_3_verify_linear_constraint(X_train, T_train, a_vec, b_const, X_test)

print(f"訓練データの制約誤差最大値: {train_viol:.2e}")
print(f"テスト予測値の制約誤差最大値: {test_viol:.2e}")

assert train_viol < 1e-12
assert test_viol < 1e-12
print("Exercise 5.3: 検証成功！")


訓練データの制約誤差最大値: 1.78e-15
テスト予測値の制約誤差最大値: 5.33e-15
Exercise 5.3: 検証成功！


---
### Exercise 5.4: 複数の同時線形制約の保存
**Problem 5.4 (? ?)**:
> Extend the result of Exercise 5.3 to show that if multiple linear constraints are satisfied simultaneously by the target vectors, then the same constraints will also be satisfied by the least-squares prediction of a linear model.

#### 数式導出と論理ステップ:
1. **複数同時線形制約の行列表現**:
   目標ベクトル $\mathbf{t}_n$ が $L$ 個の独立な線形制約を同時に満たすとします：
   $$
   \mathbf{A}^T \mathbf{t}_n + \mathbf{b} = \mathbf{0}_L, \quad \forall n \in \{1, \dots, N\}
   $$
   ここで $\mathbf{A} \in \mathbb{R}^{K \times L}$ であり、$\mathbf{b} \in \mathbb{R}^L$ です。
   各行を転置して全訓練データ $N$ 個をまとめると：
   $$
   \mathbf{t}_n^T \mathbf{A} + \mathbf{b}^T = \mathbf{0}^T \iff \mathbf{T} \mathbf{A} + \mathbf{1}_N \mathbf{b}^T = \mathbf{O}_{N \times L} \iff \mathbf{T} \mathbf{A} = -\mathbf{1}_N \mathbf{b}^T
   $$

2. **重み行列 $\mathbf{W}$ への作用**:
   最小二乗解 $\mathbf{W} = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T \mathbf{T}$ に右から制約行列 $\mathbf{A}$ を掛けます：
   $$
   \mathbf{W} \mathbf{A} = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T (\mathbf{T} \mathbf{A}) = (\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} \boldsymbol{\Phi}^T (-\mathbf{1}_N \mathbf{b}^T)
   $$
   同様に $\mathbf{1}_N = \boldsymbol{\Phi} \mathbf{e}_1$ を代入すると：
   $$
   \mathbf{W} \mathbf{A} = -(\boldsymbol{\Phi}^T \boldsymbol{\Phi})^{-1} (\boldsymbol{\Phi}^T \boldsymbol{\Phi}) \mathbf{e}_1 \mathbf{b}^T = -\mathbf{e}_1 \mathbf{b}^T \in \mathbb{R}^{M \times L}
   $$

3. **任意のテスト入力 $\mathbf{x}$ に対する予測値**:
   モデルの予測行列は $\mathbf{y}(\mathbf{x}) = \mathbf{W}^T \boldsymbol{\phi}(\mathbf{x})$ です。制約行列 $\mathbf{A}$ を作用させると：
   $$
   \mathbf{A}^T \mathbf{y}(\mathbf{x}) = (\mathbf{y}(\mathbf{x})^T \mathbf{A})^T = (\boldsymbol{\phi}(\mathbf{x})^T \mathbf{W} \mathbf{A})^T = (\boldsymbol{\phi}(\mathbf{x})^T (-\mathbf{e}_1 \mathbf{b}^T))^T = (-\phi_0(\mathbf{x}) \mathbf{b}^T)^T = -\mathbf{b}
   $$
   したがって：
   $$
   \mathbf{A}^T \mathbf{y}(\mathbf{x}) + \mathbf{b} = \mathbf{0}_L
   $$
   がすべての $\mathbf{x}$ について恒等的に成立します。特に 1-of-$K$ 分類において $\sum_k t_{nk} = 1$（すなわち $\mathbf{1}_K^T \mathbf{t}_n - 1 = 0$）という制約がある場合、最小二乗法の出力の和も必ず $\sum_k y_k(\mathbf{x}) = 1$ を満たすことが直ちに導かれます。 $\blacksquare$


In [5]:
# Exercise 5.4 数値検証
np.random.seed(42)
N, D, K, L = 60, 2, 5, 2
X_train = np.random.randn(N, D)
A_mat = np.random.randn(K, L)
b_vec = np.random.randn(L)

# 制約を満たす訓練目標値行列 T_train を射影により生成
T_raw = np.random.randn(N, K)
T_train = np.zeros_like(T_raw)
for n in range(N):
    T_train[n] = np.linalg.lstsq(A_mat.T, -b_vec, rcond=None)[0] + (
        np.eye(K) - np.linalg.pinv(A_mat.T) @ A_mat.T
    ) @ T_raw[n]

X_test = np.random.randn(25, D)
tr_v, te_v = exercise_5_4_verify_multiple_constraints(X_train, T_train, A_mat, b_vec, X_test)

print(f"訓練データの制約違反ノルム最大値: {tr_v:.2e}")
print(f"テスト予測値の制約違反ノルム最大値: {te_v:.2e}")

assert tr_v < 1e-12
assert te_v < 1e-12
print("Exercise 5.4: 検証成功！")


訓練データの制約違反ノルム最大値: 2.22e-15
テスト予測値の制約違反ノルム最大値: 1.60e-15
Exercise 5.4: 検証成功！


## Part 2: 5.2節 決定理論 (Decision Theory for Classification)


---
### Exercise 5.5: Fスコアの公式導出
**Problem 5.5 (?)**:
> Use the definition (5.38), along with (5.30) and (5.31) to derive the result (5.39) for the F-score.

#### 数式導出と論理ステップ:
1. **適合率 (Precision) と再現率 (Recall) の定義**:
   混同行列の要素（真陽性 $\text{TP}$、偽陽性 $\text{FP}$、偽陰性 $\text{FN}$）を用いて：
   $$
   P = \frac{\text{TP}}{\text{TP} + \text{FP}} \tag{5.30}, \quad R = \frac{\text{TP}}{\text{TP} + \text{FN}} \tag{5.31}
   $$

2. **重み付き調和平均による $F_\beta$ の定義 (式 5.38)**:
   適合率と再現率の重み付き調和平均として $F_\beta$ スコアは定義されます：
   $$
   \frac{1}{F_\beta} = \alpha \frac{1}{P} + (1 - \alpha) \frac{1}{R}, \quad \text{ただし } \alpha = \frac{1}{1 + \beta^2}, \; 1 - \alpha = \frac{\beta^2}{1 + \beta^2}
   $$
   両辺の逆数をとると：
   $$
   F_\beta = \frac{1}{\frac{1}{1 + \beta^2}\frac{1}{P} + \frac{\beta^2}{1 + \beta^2}\frac{1}{R}} = \frac{1 + \beta^2}{\frac{1}{P} + \frac{\beta^2}{R}} = \frac{(1 + \beta^2) P R}{\beta^2 P + R}
   $$

3. **混同行列パラメータへの代入**:
   式 (5.30) と (5.31) を代入します：
   $$
   F_\beta = \frac{(1 + \beta^2) \left(\frac{\text{TP}}{\text{TP} + \text{FP}}\right) \left(\frac{\text{TP}}{\text{TP} + \text{FN}}\right)}{\beta^2 \left(\frac{\text{TP}}{\text{TP} + \text{FP}}\right) + \left(\frac{\text{TP}}{\text{TP} + \text{FN}}\right)}
   $$
   分母と分子に $(\text{TP} + \text{FP})(\text{TP} + \text{FN})$ を掛けると：
   $$
   F_\beta = \frac{(1 + \beta^2) \text{TP}^2}{\beta^2 \text{TP}(\text{TP} + \text{FN}) + \text{TP}(\text{TP} + \text{FP})}
   $$
   分母・分子を共通因数 $\text{TP} > 0$ で割ります：
   $$
   F_\beta = \frac{(1 + \beta^2) \text{TP}}{\beta^2(\text{TP} + \text{FN}) + (\text{TP} + \text{FP})} = \frac{(1 + \beta^2) \text{TP}}{(1 + \beta^2)\text{TP} + \beta^2 \text{FN} + \text{FP}} \tag{5.39}
   $$
   これにより、式 (5.39) が完全に導出されました。 $\blacksquare$


In [6]:
# Exercise 5.5 数値検証
TP, FP, FN = 85, 15, 25
for beta in [0.5, 1.0, 2.0]:
    f_pr, f_form = exercise_5_5_compute_f_score(TP, FP, FN, beta)
    print(f"beta = {beta:3.1f}: F_score (P, R定義) = {f_pr:.6f}, F_score (式 5.39) = {f_form:.6f}")
    assert np.isclose(f_pr, f_form)

print("Exercise 5.5: 検証成功！")


beta = 0.5: F_score (P, R定義) = 0.833333, F_score (式 5.39) = 0.833333
beta = 1.0: F_score (P, R定義) = 0.809524, F_score (式 5.39) = 0.809524
beta = 2.0: F_score (P, R定義) = 0.787037, F_score (式 5.39) = 0.787037
Exercise 5.5: 検証成功！


---
### Exercise 5.6: バタチャリア誤分類確率上界の導出
**Problem 5.6 (? ?)**:
> Consider two non-negative numbers $a$ and $b$, and show that, if $a \le b$, then $a \le (ab)^{1/2}$. Use this result to show that, if the decision regions of a two-class classification problem are chosen to minimize the probability of misclassification, this probability will satisfy $p(\text{mistake}) \le \int \{p(\mathbf{x}, \mathcal{C}_1)p(\mathbf{x}, \mathcal{C}_2)\}^{1/2} d\mathbf{x}$.

#### 数式導出と論理ステップ:
1. **不等式 $a \le (ab)^{1/2}$ の証明**:
   $a, b \ge 0$ かつ $a \le b$ であるとき、両辺に非負の $a$ を掛けると：
   $$
   a^2 \le ab
   $$
   平方根をとると（両辺とも非負であるため）：
   $$
   a \le (ab)^{1/2}
   $$
   一般に、任意の非負の実数 $a, b \ge 0$ に対して $\min(a, b) \le a$ かつ $\min(a, b) \le b$ ですので：
   $$
   \min(a, b)^2 \le ab \implies \min(a, b) \le \sqrt{ab}
   $$
   が常に成り立ちます。

2. **誤分類確率の定式化**:
   2クラス分類において、誤分類確率を最小化するベイズ決定規則は：
   $$
   \mathcal{R}_1 = \{\mathbf{x} \mid p(\mathbf{x}, \mathcal{C}_1) \ge p(\mathbf{x}, \mathcal{C}_2)\}, \quad \mathcal{R}_2 = \{\mathbf{x} \mid p(\mathbf{x}, \mathcal{C}_2) > p(\mathbf{x}, \mathcal{C}_1)\}
   $$
   です。このとき誤分類確率は次のように積分されます：
   $$
   p(\text{mistake}) = \int_{\mathcal{R}_1} p(\mathbf{x}, \mathcal{C}_2) d\mathbf{x} + \int_{\mathcal{R}_2} p(\mathbf{x}, \mathcal{C}_1) d\mathbf{x} = \int \min\{p(\mathbf{x}, \mathcal{C}_1), \; p(\mathbf{x}, \mathcal{C}_2)\} d\mathbf{x}
   $$

3. **バタチャリア上界の適用**:
   各点 $\mathbf{x}$ において $\min\{p(\mathbf{x}, \mathcal{C}_1), p(\mathbf{x}, \mathcal{C}_2)\} \le \{p(\mathbf{x}, \mathcal{C}_1) p(\mathbf{x}, \mathcal{C}_2)\}^{1/2}$ が成立するため：
   $$
   p(\text{mistake}) \le \int \{p(\mathbf{x}, \mathcal{C}_1) p(\mathbf{x}, \mathcal{C}_2)\}^{1/2} d\mathbf{x} \tag{5.99}
   $$
   が示されました。この上界はバタチャリア係数 (Bhattacharyya coefficient) として知られ、ガウス分布などの下で閉形式で評価可能な有用な理論的限界を与えます。 $\blacksquare$


In [7]:
# Exercise 5.6 数値検証
# 1次元ガウス分布による2クラス問題
p_j1 = lambda x: 0.5 * norm.pdf(x, loc=-1.0, scale=1.0)
p_j2 = lambda x: 0.5 * norm.pdf(x, loc=1.0, scale=1.2)

err_true, err_bound = exercise_5_6_compute_bhattacharyya_bound(p_j1, p_j2)

print(f"真の最小誤分類確率 p(mistake):   {err_true:.6f}")
print(f"バタチャリア理論上界 (式 5.99):   {err_bound:.6f}")

assert err_true <= err_bound, "誤分類確率はバタチャリア上界以下でなければなりません。"
print("Exercise 5.6: 検証成功！")


真の最小誤分類確率 p(mistake):   0.180461
バタチャリア理論上界 (式 5.99):   0.329148
Exercise 5.6: 検証成功！


---
### Exercise 5.7: 0-1 損失行列と事後確率最大化の等価性
**Problem 5.7 (?)**:
> Given a loss matrix with elements $L_{kj}$, the expected risk is minimized if, for each $\mathbf{x}$, we choose the class that minimizes (5.23). Verify that, when the loss matrix is given by $L_{kj} = 1 - I_{kj}$, where $I_{kj}$ are the elements of the identity matrix, this reduces to the criterion of choosing the class having the largest posterior probability. What is the interpretation of this form of loss matrix?

#### 数式導出と論理ステップ:
1. **期待損失 (式 5.23) の定式化**:
   入力 $\mathbf{x}$ に対してクラス $\mathcal{C}_j$ を予測したときの条件付き期待損失は：
   $$
   \mathbb{E}[L_j \mid \mathbf{x}] = \sum_{k=1}^K L_{kj} \, p(\mathcal{C}_k \mid \mathbf{x}) \tag{5.23}
   $$

2. **0-1 損失行列 $L_{kj} = 1 - I_{kj}$ の代入**:
   正解が $k$ で予測が $j$ のとき、$k = j$（正解）ならば損失 0、$k \ne j$（不正解）ならば損失 1 を与える 0-1 損失行列を代入します：
   $$
   \mathbb{E}[L_j \mid \mathbf{x}] = \sum_{k=1}^K (1 - I_{kj}) p(\mathcal{C}_k \mid \mathbf{x}) = \sum_{k=1}^K p(\mathcal{C}_k \mid \mathbf{x}) - \sum_{k=1}^K I_{kj} p(\mathcal{C}_k \mid \mathbf{x})
   $$
   確率の規格化条件 $\sum_{k=1}^K p(\mathcal{C}_k \mid \mathbf{x}) = 1$ より：
   $$
   \mathbb{E}[L_j \mid \mathbf{x}] = 1 - p(\mathcal{C}_j \mid \mathbf{x})
   $$

3. **損失最小化基準の同値性**:
   期待損失を最小化する最適決定 $j^*$ は：
   $$
   j^* = \arg\min_j \mathbb{E}[L_j \mid \mathbf{x}] = \arg\min_j \{1 - p(\mathcal{C}_j \mid \mathbf{x})\} = \arg\max_j p(\mathcal{C}_j \mid \mathbf{x})
   $$
   となり、最大の事後確率を持つクラスを選択する MAP (Maximum A Posteriori) 規則と完全に一致します。

4. **この損失行列の解釈**:
   正解（$k=j$）にはペナルティを与えず（損失 0）、どのような種類の誤分類（例えば良性を悪性と誤診する場合と、悪性を良性と誤診する場合）であっても全く対等にペナルティ 1 を課す対称的な誤分類コストを表します。 $\blacksquare$


In [8]:
# Exercise 5.7 数値検証
p_post = np.array([0.1, 0.65, 0.25])
best_class, losses = exercise_5_7_zero_one_loss_decision(p_post)

print("事後確率:", p_post)
print("各クラスを選択したときの期待損失 1 - p(C_j|x):", losses)
print("最小損失クラス:", best_class, "最大事後確率クラス:", np.argmax(p_post))

assert best_class == np.argmax(p_post)
assert np.allclose(losses, 1.0 - p_post)
print("Exercise 5.7: 検証成功！")


事後確率: [0.1  0.65 0.25]
各クラスを選択したときの期待損失 1 - p(C_j|x): [0.9  0.35 0.75]
最小損失クラス: 1 最大事後確率クラス: 1
Exercise 5.7: 検証成功！


---
### Exercise 5.8: 一般損失行列と事前確率の下での期待損失最小化基準
**Problem 5.8 (?)**:
> Derive the criterion for minimizing the expected loss when there is a general loss matrix and general prior probabilities for the classes.

#### 数式導出と論理ステップ:
1. **総期待損失の定式化**:
   空間全体を各決定領域 $\mathcal{R}_j$（クラス $\mathcal{C}_j$ に割り当てる領域）に分割したときの総期待損失（リスク）は：
   $$
   \mathbb{E}[L] = \sum_{k=1}^K \sum_{j=1}^K \int_{\mathcal{R}_j} L_{kj} \, p(\mathbf{x}, \mathcal{C}_k) \, d\mathbf{x} = \sum_{j=1}^K \int_{\mathcal{R}_j} \left\{ \sum_{k=1}^K L_{kj} \, p(\mathbf{x}, \mathcal{C}_k) \right\} d\mathbf{x}
   $$

2. **ベイズの定理による結合確率の展開**:
   結合確率は $p(\mathbf{x}, \mathcal{C}_k) = p(\mathbf{x} \mid \mathcal{C}_k) p(\mathcal{C}_k) = p(\mathcal{C}_k \mid \mathbf{x}) p(\mathbf{x})$ と書けます。したがって：
   $$
   \mathbb{E}[L] = \int \left( \sum_{j=1}^K I(\mathbf{x} \in \mathcal{R}_j) \sum_{k=1}^K L_{kj} p(\mathbf{x} \mid \mathcal{C}_k) p(\mathcal{C}_k) \right) d\mathbf{x}
   $$

3. **各点 $\mathbf{x}$ での最小化決定基準**:
   総損失 $\mathbb{E}[L]$ を最小化するためには、各入力点 $\mathbf{x}$ に対して被積分関数を最小化する領域 $\mathcal{R}_j$ に割り当てればよいです：
   $$
   \mathbf{x} \in \mathcal{R}_j \iff \sum_{k=1}^K L_{kj} \, p(\mathbf{x} \mid \mathcal{C}_k) p(\mathcal{C}_k) \le \sum_{k=1}^K L_{ki} \, p(\mathbf{x} \mid \mathcal{C}_k) p(\mathcal{C}_k) \quad (\forall i \ne j)
   $$
   事後確率を用いて表せば、$p(\mathbf{x})$ で割ることで：
   $$
   j^*(\mathbf{x}) = \arg\min_j \sum_{k=1}^K L_{kj} \, p(\mathcal{C}_k \mid \mathbf{x})
   $$
   が得られます。 $\blacksquare$


In [9]:
# Exercise 5.8 数値検証
# 非対称な損失行列（重篤な疾患の見逃しに重いペナルティ）
L_matrix = np.array([
    [0.0, 10.0],  # 実際は健康(C0): 予測健康なら0, 誤診なら10
    [100.0, 0.0]  # 実際は重症(C1): 予測健康(見逃し)なら100, 予測重症なら0
])
p_post_patient = np.array([0.88, 0.12])  # 罹患率はわずか12%だが...

chosen_act, exp_losses = exercise_5_8_expected_loss_decision(L_matrix, p_post_patient)
print("各行動をとった場合の期待損失 [予測健康, 予測重症]:", exp_losses)
print("最適な行動 (0: 診断健康, 1: 診断重症/精密検査):", chosen_act)

# 事後確率は健康(88%)が高いが、期待損失は重症と判定したほうが小さい(8.8 < 12.0)
assert chosen_act == 1
print("Exercise 5.8: 検証成功！")


各行動をとった場合の期待損失 [予測健康, 予測重症]: [12.   8.8]
最適な行動 (0: 診断健康, 1: 診断重症/精密検査): 1
Exercise 5.8: 検証成功！


---
### Exercise 5.9: 事後確率の標本平均の事前確率への漸近収束
**Problem 5.9 (?)**:
> Consider the average of the posterior probabilities over a set of $N$ data points in the form $\frac{1}{N} \sum_{n=1}^N p(\mathcal{C}_k \mid \mathbf{x}_n)$. By taking the limit $N \to \infty$, show that this quantity approaches the prior class probability $p(\mathcal{C}_k)$.

#### 数式導出と論理ステップ:
1. **大数の法則 (Law of Large Numbers) の適用**:
   訓練データ点 $\mathbf{x}_n$ ($n = 1, \dots, N$) がデータ生成分布 $p(\mathbf{x})$ から独立同分布 (i.i.d.) にサンプリングされていると仮定します。
   このとき、任意の可積分関数 $f(\mathbf{x})$ の標本平均は大数の強法則により真の期待値へ概収束します：
   $$
   \lim_{N \to \infty} \frac{1}{N} \sum_{n=1}^N f(\mathbf{x}_n) = \mathbb{E}_{\mathbf{x}}[f(\mathbf{x})] = \int f(\mathbf{x}) p(\mathbf{x}) d\mathbf{x}
   $$

2. **事後確率の期待値の計算**:
   関数として事後確率 $f(\mathbf{x}) = p(\mathcal{C}_k \mid \mathbf{x})$ を代入します：
   $$
   \lim_{N \to \infty} \frac{1}{N} \sum_{n=1}^N p(\mathcal{C}_k \mid \mathbf{x}_n) = \int p(\mathcal{C}_k \mid \mathbf{x}) p(\mathbf{x}) d\mathbf{x}
   $$

3. **周辺化による事前確率の導出**:
   条件付き確率の定義 $p(\mathcal{C}_k \mid \mathbf{x}) p(\mathbf{x}) = p(\mathbf{x}, \mathcal{C}_k)$ を代入し、入力空間全体で周辺化（積分）します：
   $$
   \int p(\mathcal{C}_k \mid \mathbf{x}) p(\mathbf{x}) d\mathbf{x} = \int p(\mathbf{x}, \mathcal{C}_k) d\mathbf{x} = p(\mathcal{C}_k)
   $$
   したがって：
   $$
   \lim_{N \to \infty} \frac{1}{N} \sum_{n=1}^N p(\mathcal{C}_k \mid \mathbf{x}_n) = p(\mathcal{C}_k) \tag{5.100}
   $$
   が示されました。 $\blacksquare$


In [10]:
# Exercise 5.9 数値検証（モンテカルロシミュレーション）
np.random.seed(42)
N_samples = 30000
true_prior_C1 = 0.65

# 事前確率からクラス生成
classes = (np.random.rand(N_samples) < true_prior_C1).astype(int)
X_samples = np.where(classes == 1, np.random.randn(N_samples) + 1.0, np.random.randn(N_samples) - 1.0)

# 真の事後確率 p(C1|x)
p_x_c1 = true_prior_C1 * norm.pdf(X_samples, 1.0, 1.0)
p_x_c0 = (1.0 - true_prior_C1) * norm.pdf(X_samples, -1.0, 1.0)
p_post_C1 = p_x_c1 / (p_x_c1 + p_x_c0)

avg_posterior = exercise_5_9_average_posterior(p_post_C1)
print(f"真の事前確率 p(C1): {true_prior_C1:.4f}")
print(f"事後確率の標本平均 (N={N_samples}): {avg_posterior:.4f}")

assert np.isclose(avg_posterior, true_prior_C1, atol=0.01)
print("Exercise 5.9: 検証成功！")


真の事前確率 p(C1): 0.6500
事後確率の標本平均 (N=30000): 0.6519
Exercise 5.9: 検証成功！


---
### Exercise 5.10: 棄却オプション付き決定基準と閾値関係式
**Problem 5.10 (? ?)**:
> Consider a classification problem in which the loss incurred when an input vector from class $\mathcal{C}_k$ is classified as belonging to class $\mathcal{C}_j$ is given by the loss matrix $L_{kj}$ and for which the loss incurred in selecting the reject option is $\lambda$. Find the decision criterion that will give the minimum expected loss. Verify that this reduces to the reject criterion discussed in Section 5.2.3 when the loss matrix is given by $L_{kj} = 1 - I_{kj}$. What is the relationship between $\lambda$ and the rejection threshold $\theta$?

#### 数式導出と論理ステップ:
1. **棄却オプションを含む決定問題の定式化**:
   意思決定者は各入力 $\mathbf{x}$ に対して、$K$ 個のクラス $\mathcal{C}_1, \dots, \mathcal{C}_K$ のいずれかに分類するか、あるいは判断を保留して「棄却 (Reject, $\text{action } R$)」を選択できます。
   各決定に対する条件付き期待損失は：
   $$
   \mathbb{E}[\text{Loss} \mid \text{choose } j, \mathbf{x}] = \sum_{k=1}^K L_{kj} \, p(\mathcal{C}_k \mid \mathbf{x}), \quad \mathbb{E}[\text{Loss} \mid \text{reject}, \mathbf{x}] = \lambda
   $$

2. **一般損失行列における最適決定規則**:
   期待損失を最小化するためには、最小の損失をもたらす選択肢を採択します：
   $$
   \begin{cases}
   \text{クラス } \mathcal{C}_j \text{ に分類} & \text{if } \min_i \sum_{k=1}^K L_{ki} p(\mathcal{C}_k \mid \mathbf{x}) = \sum_{k=1}^K L_{kj} p(\mathcal{C}_k \mid \mathbf{x}) \le \lambda \\
   \text{棄却 (Reject)} & \text{if } \min_i \sum_{k=1}^K L_{ki} p(\mathcal{C}_k \mid \mathbf{x}) > \lambda
   \end{cases}
   $$

3. **0-1 損失行列 $L_{kj} = 1 - I_{kj}$ への適用**:
   0-1 損失のとき、クラス $\mathcal{C}_j$ を選択したときの期待損失は Exercise 5.7 より $1 - p(\mathcal{C}_j \mid \mathbf{x})$ です。
   全クラスにわたる最小損失は：
   $$
   \min_i \sum_{k=1}^K L_{ki} p(\mathcal{C}_k \mid \mathbf{x}) = \min_i \{1 - p(\mathcal{C}_i \mid \mathbf{x})\} = 1 - \max_k p(\mathcal{C}_k \mid \mathbf{x})
   $$
   分類を行う条件は：
   $$
   1 - \max_k p(\mathcal{C}_k \mid \mathbf{x}) \le \lambda \iff \max_k p(\mathcal{C}_k \mid \mathbf{x}) \ge 1 - \lambda
   $$
   逆に棄却を行う条件は：
   $$
   \max_k p(\mathcal{C}_k \mid \mathbf{x}) < 1 - \lambda
   $$

4. **棄却閾値 $\theta$ との厳密な関係**:
   5.2.3節の棄却基準「最大事後確率が閾値 $\theta$ 未満のときに棄却する」と比較すると：
   $$
   \theta = 1 - \lambda
   $$
   という直接的な等式関係が成立します。
   - $\lambda = 0$（棄却コストなし）ならば $\theta = 1$ となり、100% 確信がない限りすべて棄却されます。
   - $\lambda \ge 1$（棄却コストが誤分類損失以上）ならば $\theta \le 0$ となり、棄却は一切選ばれず常に分類されます。 $\blacksquare$


In [11]:
# Exercise 5.10 数値検証
for lambda_cost in [0.1, 0.25, 0.4]:
    theta = exercise_5_10_reject_threshold(lambda_cost)
    print(f"棄却コスト lambda = {lambda_cost:.2f} => 棄却閾値 theta = {theta:.2f}")
    assert np.isclose(theta, 1.0 - lambda_cost)

print("Exercise 5.10: 検証成功！")


棄却コスト lambda = 0.10 => 棄却閾値 theta = 0.90
棄却コスト lambda = 0.25 => 棄却閾値 theta = 0.75
棄却コスト lambda = 0.40 => 棄却閾値 theta = 0.60
Exercise 5.10: 検証成功！


## Part 3: 5.3節 生成モデル (Generative Classifiers)


---
### Exercise 5.11: ロジスティック・シグモイド関数の対称性と逆関数
**Problem 5.11 (?)**:
> Show that the logistic sigmoid function (5.42) satisfies the property $\sigma(-a) = 1 - \sigma(a)$ and that its inverse is given by $\sigma^{-1}(y) = \ln \{y/(1 - y)\}$.

#### 数式導出と論理ステップ:
1. **対称性 $\sigma(-a) = 1 - \sigma(a)$ の証明**:
   ロジスティック・シグモイド関数の定義 (式 5.42) より：
   $$
   \sigma(a) = \frac{1}{1 + e^{-a}}
   $$
   入力に $-a$ を代入します：
   $$
   \sigma(-a) = \frac{1}{1 + e^a}
   $$
   分母・分子に $e^{-a}$ を掛けると：
   $$
   \sigma(-a) = \frac{e^{-a}}{e^{-a}(1 + e^a)} = \frac{e^{-a}}{e^{-a} + 1} = \frac{(1 + e^{-a}) - 1}{1 + e^{-a}} = 1 - \frac{1}{1 + e^{-a}} = 1 - \sigma(a)
   $$
   が示されました。

2. **逆関数ロジット (logit) の導出**:
   $y = \sigma(a) = \frac{1}{1 + e^{-a}}$ とおき、$a$ について解きます（ただし $y \in (0, 1)$）：
   $$
   1 + e^{-a} = \frac{1}{y} \implies e^{-a} = \frac{1}{y} - 1 = \frac{1 - y}{y}
   $$
   両辺の逆数をとると：
   $$
   e^a = \frac{y}{1 - y}
   $$
   両辺の自然対数をとると：
   $$
   a = \ln \left\{ \frac{y}{1 - y} \right\} = \operatorname{logit}(y)
   $$
   となり、$\sigma^{-1}(y) = \ln \{y / (1 - y)\}$ が証明されました。 $\blacksquare$


In [12]:
# Exercise 5.11 数値検証
a_grid = np.linspace(-5.0, 5.0, 50)
sym_ok, inv_ok = exercise_5_11_verify_sigmoid_properties(a_grid)

print("対称性 sigma(-a) == 1 - sigma(a) の検証:", sym_ok)
print("逆関数 logit(sigma(a)) == a の検証:", inv_ok)

assert sym_ok and inv_ok
print("Exercise 5.11: 検証成功！")


対称性 sigma(-a) == 1 - sigma(a) の検証: True
逆関数 logit(sigma(a)) == a の検証: True
Exercise 5.11: 検証成功！


---
### Exercise 5.12: 2クラス共通共分散ガウス生成モデルの事後確率パラメータ
**Problem 5.12 (?)**:
> Using (5.40) and (5.41), derive the result (5.48) for the posterior class probability in the two-class generative model with Gaussian densities, and verify the results (5.49) and (5.50) for the parameters $\mathbf{w}$ and $w_0$.

#### 数式導出と論理ステップ:
1. **事後確率のロジスティック・シグモイド表現 (式 5.40 - 5.41)**:
   ベイズの定理より：
   $$
   p(\mathcal{C}_1 \mid \mathbf{x}) = \frac{p(\mathbf{x} \mid \mathcal{C}_1) p(\mathcal{C}_1)}{p(\mathbf{x} \mid \mathcal{C}_1) p(\mathcal{C}_1) + p(\mathbf{x} \mid \mathcal{C}_2) p(\mathcal{C}_2)} = \frac{1}{1 + \exp(-a)} = \sigma(a)
   $$
   ここで事前対数オッズを含む量 $a$ は：
   $$
   a = \ln \frac{p(\mathbf{x} \mid \mathcal{C}_1) p(\mathcal{C}_1)}{p(\mathbf{x} \mid \mathcal{C}_2) p(\mathcal{C}_2)} = \ln p(\mathbf{x} \mid \mathcal{C}_1) - \ln p(\mathbf{x} \mid \mathcal{C}_2) + \ln \frac{p(\mathcal{C}_1)}{p(\mathcal{C}_2)}
   $$

2. **共通共分散多変量ガウス密度の代入**:
   $p(\mathbf{x} \mid \mathcal{C}_k) = \mathcal{N}(\mathbf{x} \mid \boldsymbol{\mu}_k, \boldsymbol{\Sigma})$ の対数は：
   $$
   \ln p(\mathbf{x} \mid \mathcal{C}_k) = -\frac{D}{2}\ln(2\pi) - \frac{1}{2}\ln|\boldsymbol{\Sigma}| - \frac{1}{2}(\mathbf{x} - \boldsymbol{\mu}_k)^T \boldsymbol{\Sigma}^{-1}(\mathbf{x} - \boldsymbol{\mu}_k)
   $$
   差分をとると定数項と正規化係数は相殺されます：
   $$
   \ln p(\mathbf{x} \mid \mathcal{C}_1) - \ln p(\mathbf{x} \mid \mathcal{C}_2) = -\frac{1}{2}(\mathbf{x} - \boldsymbol{\mu}_1)^T \boldsymbol{\Sigma}^{-1}(\mathbf{x} - \boldsymbol{\mu}_1) + \frac{1}{2}(\mathbf{x} - \boldsymbol{\mu}_2)^T \boldsymbol{\Sigma}^{-1}(\mathbf{x} - \boldsymbol{\mu}_2)
   $$

3. **二次項の相殺と線形パラメータの同定**:
   内積を展開します：
   $$
   (\mathbf{x} - \boldsymbol{\mu}_k)^T \boldsymbol{\Sigma}^{-1}(\mathbf{x} - \boldsymbol{\mu}_k) = \mathbf{x}^T \boldsymbol{\Sigma}^{-1}\mathbf{x} - 2 \boldsymbol{\mu}_k^T \boldsymbol{\Sigma}^{-1}\mathbf{x} + \boldsymbol{\mu}_k^T \boldsymbol{\Sigma}^{-1}\boldsymbol{\mu}_k
   $$
   共分散行列が共通であるため、2次の項 $\mathbf{x}^T \boldsymbol{\Sigma}^{-1}\mathbf{x}$ が完全に相殺されます：
   $$
   a = (\boldsymbol{\mu}_1 - \boldsymbol{\mu}_2)^T \boldsymbol{\Sigma}^{-1}\mathbf{x} - \frac{1}{2}\boldsymbol{\mu}_1^T \boldsymbol{\Sigma}^{-1}\boldsymbol{\mu}_1 + \frac{1}{2}\boldsymbol{\mu}_2^T \boldsymbol{\Sigma}^{-1}\boldsymbol{\mu}_2 + \ln \frac{p(\mathcal{C}_1)}{p(\mathcal{C}_2)}
   $$
   これを線形形式 $a = \mathbf{w}^T \mathbf{x} + w_0$ と比較することにより：
   $$
   \mathbf{w} = \boldsymbol{\Sigma}^{-1}(\boldsymbol{\mu}_1 - \boldsymbol{\mu}_2) \tag{5.49}
   $$
   $$
   w_0 = -\frac{1}{2}\boldsymbol{\mu}_1^T \boldsymbol{\Sigma}^{-1}\boldsymbol{\mu}_1 + \frac{1}{2}\boldsymbol{\mu}_2^T \boldsymbol{\Sigma}^{-1}\boldsymbol{\mu}_2 + \ln \frac{p(\mathcal{C}_1)}{p(\mathcal{C}_2)} \tag{5.50}
   $$
   が導かれました。 $\blacksquare$


In [13]:
# Exercise 5.12 数値検証
mu1 = np.array([2.0, 1.0])
mu2 = np.array([-1.0, -1.0])
Sigma = np.array([[1.5, 0.4], [0.4, 1.2]])
pi1, pi2 = 0.6, 0.4

w, w0 = exercise_5_12_compute_gaussian_gda_params(mu1, mu2, Sigma, pi1, pi2)

# テスト点におけるベイズ事後確率とシグモイド値の直接比較
x_test = np.array([0.5, 0.2])
a_lin = np.dot(w, x_test) + w0
prob_sigmoid = sigmoid(a_lin)

inv_cov = np.linalg.inv(Sigma)
norm1 = np.exp(-0.5 * (x_test - mu1).T @ inv_cov @ (x_test - mu1))
norm2 = np.exp(-0.5 * (x_test - mu2).T @ inv_cov @ (x_test - mu2))
prob_bayes = (pi1 * norm1) / (pi1 * norm1 + pi2 * norm2)

print(f"シグモイド予測事後確率: {prob_sigmoid:.6f}")
print(f"直接ベイズ計算事後確率: {prob_bayes:.6f}")

assert np.isclose(prob_sigmoid, prob_bayes)
print("Exercise 5.12: 検証成功！")


シグモイド予測事後確率: 0.651350
直接ベイズ計算事後確率: 0.651350
Exercise 5.12: 検証成功！


---
### Exercise 5.13: 多クラス事前確率のラグランジュ最尤推定
**Problem 5.13 (?)**:
> Consider a generative classification model for $K$ classes defined by prior class probabilities $p(\mathcal{C}_k) = \pi_k$ and general class-conditional densities $p(\boldsymbol{\phi} \mid \mathcal{C}_k)$. Show that the maximum-likelihood solution for the prior probabilities is given by $\pi_k = N_k / N$.

#### 数式導出と論理ステップ:
1. **尤度関数の定式化**:
   訓練データ $\{\boldsymbol{\phi}_n, \mathbf{t}_n\}_{n=1}^N$ が独立同分布に生成されたとします。
   1-of-$K$ 符号化目標値ベクトル $\mathbf{t}_n$ を用いると、結合対数尤度関数は：
   $$
   \ln p(\mathbf{T}, \boldsymbol{\Phi} \mid \boldsymbol{\pi}, \boldsymbol{\theta}) = \sum_{n=1}^N \sum_{k=1}^K t_{nk} \left\{ \ln \pi_k + \ln p(\boldsymbol{\phi}_n \mid \mathcal{C}_k, \boldsymbol{\theta}_k) \right\}
   $$
   事前確率 $\boldsymbol{\pi} = (\pi_1, \dots, \pi_K)^T$ に依存する部分は：
   $$
   \ln p(\mathbf{T} \mid \boldsymbol{\pi}) = \sum_{k=1}^K \left(\sum_{n=1}^N t_{nk}\right) \ln \pi_k = \sum_{k=1}^K N_k \ln \pi_k
   $$
   ここで $N_k = \sum_{n=1}^N t_{nk}$ はクラス $\mathcal{C}_k$ に属するデータ数です。

2. **制約付き最適化問題とラグランジアン**:
   事前確率は $\sum_{k=1}^K \pi_k = 1$ を満たさなければなりません。ラグランジュ未定乗数 $\lambda$ を導入します：
   $$
   \mathcal{L}(\boldsymbol{\pi}, \lambda) = \sum_{k=1}^K N_k \ln \pi_k + \lambda \left( \sum_{k=1}^K \pi_k - 1 \right)
   $$

3. **停留条件と乗数の決定**:
   各 $\pi_k$ に関して偏微分して 0 とおきます：
   $$
   \frac{\partial \mathcal{L}}{\partial \pi_k} = \frac{N_k}{\pi_k} + \lambda = 0 \implies \pi_k = -\frac{N_k}{\lambda}
   $$
   制約条件 $\sum_{k=1}^K \pi_k = 1$ に代入します：
   $$
   \sum_{k=1}^K \left(-\frac{N_k}{\lambda}\right) = -\frac{1}{\lambda} \sum_{k=1}^K N_k = -\frac{N}{\lambda} = 1 \implies \lambda = -N
   $$
   したがって：
   $$
   \pi_k = \frac{N_k}{N} \tag{5.101}
   $$
   が厳密に導出されました。 $\blacksquare$


In [14]:
# Exercise 5.13 数値検証
targets_sample = np.array([0, 0, 1, 1, 1, 2, 2, 2, 2])
K_cls = 3
pi_mle = exercise_5_13_mle_priors(targets_sample, K_cls)

expected_pi = np.array([2/9, 3/9, 4/9])
print("推定された事前確率 pi_k:", pi_mle)
print("理論期待値 N_k / N:     ", expected_pi)

assert np.allclose(pi_mle, expected_pi)
print("Exercise 5.13: 検証成功！")


推定された事前確率 pi_k: [0.22222222 0.33333333 0.44444444]
理論期待値 N_k / N:      [0.22222222 0.33333333 0.44444444]
Exercise 5.13: 検証成功！


---
### Exercise 5.14: 共通共分散ガウス生成モデルの最尤推定量
**Problem 5.14 (? ?)**:
> Show that the maximum likelihood solution for the mean of the Gaussian distribution for class $\mathcal{C}_k$ is given by (5.103), and show that the maximum likelihood solution for the shared covariance matrix is given by (5.104).

#### 数式導出と論理ステップ:
1. **対数尤度関数の定式化**:
   クラス付きガウス密度 $p(\boldsymbol{\phi} \mid \mathcal{C}_k) = \mathcal{N}(\boldsymbol{\phi} \mid \boldsymbol{\mu}_k, \boldsymbol{\Sigma})$ の結合対数尤度は：
   $$
   \ln L = -\frac{1}{2} \sum_{n=1}^N \sum_{k=1}^K t_{nk} \left\{ D \ln(2\pi) + \ln|\boldsymbol{\Sigma}| + (\boldsymbol{\phi}_n - \boldsymbol{\mu}_k)^T \boldsymbol{\Sigma}^{-1}(\boldsymbol{\phi}_n - \boldsymbol{\mu}_k) \right\}
   $$

2. **クラス平均 $\boldsymbol{\mu}_k$ の最尤解の導出**:
   $\boldsymbol{\mu}_k$ に関する項を取り出し微分します：
   $$
   \frac{\partial \ln L}{\partial \boldsymbol{\mu}_k} = \sum_{n=1}^N t_{nk} \boldsymbol{\Sigma}^{-1}(\boldsymbol{\phi}_n - \boldsymbol{\mu}_k) = \mathbf{0}
   $$
   左から正則行列 $\boldsymbol{\Sigma}$ を掛けると：
   $$
   \sum_{n=1}^N t_{nk} \boldsymbol{\phi}_n - \left( \sum_{n=1}^N t_{nk} \right) \boldsymbol{\mu}_k = \mathbf{0} \implies \boldsymbol{\mu}_k = \frac{1}{N_k} \sum_{n=1}^N t_{nk} \boldsymbol{\phi}_n \tag{5.103}
   $$
   が得られます。

3. **共通共分散行列 $\boldsymbol{\Sigma}$ の最尤解の導出**:
   精度行列 $\mathbf{K} = \boldsymbol{\Sigma}^{-1}$ で尤度を書き換えます（$\ln|\boldsymbol{\Sigma}| = -\ln|\mathbf{K}|$）：
   $$
   \ln L = \frac{N}{2}\ln|\mathbf{K}| - \frac{1}{2} \sum_{n=1}^N \sum_{k=1}^K t_{nk} \operatorname{Tr}\left[ \mathbf{K} (\boldsymbol{\phi}_n - \boldsymbol{\mu}_k)(\boldsymbol{\phi}_n - \boldsymbol{\mu}_k)^T \right]
   $$
   各クラスの標本共分散行列を行列 $\mathbf{S}_k = \frac{1}{N_k} \sum_{n=1}^N t_{nk} (\boldsymbol{\phi}_n - \boldsymbol{\mu}_k)(\boldsymbol{\phi}_n - \boldsymbol{\mu}_k)^T$ と定義すると：
   $$
   \ln L = \frac{N}{2}\ln|\mathbf{K}| - \frac{1}{2} \operatorname{Tr}\left[ \mathbf{K} \sum_{k=1}^K N_k \mathbf{S}_k \right]
   $$
   行列微分公式 $\frac{\partial \ln|\mathbf{K}|}{\partial \mathbf{K}} = \mathbf{K}^{-1}$ および $\frac{\partial \operatorname{Tr}[\mathbf{K}\mathbf{A}]}{\partial \mathbf{K}} = \mathbf{A}$ より：
   $$
   \frac{\partial \ln L}{\partial \mathbf{K}} = \frac{N}{2} \mathbf{K}^{-1} - \frac{1}{2} \sum_{k=1}^K N_k \mathbf{S}_k = \mathbf{O}
   $$
   $\mathbf{K}^{-1} = \boldsymbol{\Sigma}$ を代入して整理すると：
   $$
   \boldsymbol{\Sigma} = \sum_{k=1}^K \frac{N_k}{N} \mathbf{S}_k \tag{5.104}
   $$
   となり、各クラスの標本共分散を行積率（事前確率 $\pi_k = N_k/N$）で加重平均したものが共通共分散行列の最尤推定量であることが証明されました。 $\blacksquare$


In [15]:
# Exercise 5.14 数値検証
np.random.seed(42)
Phi_data = np.vstack([
    np.random.randn(30, 2) + np.array([2.0, 2.0]),
    np.random.randn(40, 2) + np.array([-2.0, -2.0]),
    np.random.randn(30, 2) + np.array([2.0, -2.0])
])
target_data = np.array([0]*30 + [1]*40 + [2]*30)

means_est, cov_est = exercise_5_14_mle_gaussian_shared_cov(Phi_data, target_data, 3)

print("クラス0 推定平均:", means_est[0])
print("クラス1 推定平均:", means_est[1])
print("クラス2 推定平均:", means_est[2])
print("推定共通共分散行列 Sigma:")
print(cov_est)

assert len(means_est) == 3
assert cov_est.shape == (2, 2)
assert np.all(np.linalg.eigvalsh(cov_est) > 0)
print("Exercise 5.14: 検証成功！")


クラス0 推定平均: [1.85584744 1.8348432 ]
クラス1 推定平均: [-2.11974898 -1.94236919]
クラス2 推定平均: [ 1.91860369 -1.79827653]
推定共通共分散行列 Sigma:
[[0.72533717 0.023982  ]
 [0.023982   0.96717949]]
Exercise 5.14: 検証成功！


---
### Exercise 5.15: 離散2値ベルヌーイナイーブベイズの最尤推定
**Problem 5.15 (? ?)**:
> Derive the maximum likelihood solution for the parameters $\{\mu_{ki}\}$ of the probabilistic naive Bayes classifier with discrete binary features described in Section 5.3.3.

#### 数式導出と論理ステップ:
1. **2値ナイーブベイズの尤度定式化**:
   特徴量 $\mathbf{x} = (x_1, \dots, x_D)^T \in \{0, 1\}^D$ の各成分がクラス $\mathcal{C}_k$ の下で条件付き独立にベルヌーイ分布に従うとします：
   $$
   p(\mathbf{x} \mid \mathcal{C}_k) = \prod_{i=1}^D \mu_{ki}^{x_i} (1 - \mu_{ki})^{1 - x_i}
   $$
   全訓練データに対する対数尤度関数は：
   $$
   \ln L = \sum_{n=1}^N \sum_{k=1}^K t_{nk} \sum_{i=1}^D \left\{ x_{ni} \ln \mu_{ki} + (1 - x_{ni}) \ln(1 - \mu_{ki}) \right\} + \text{const}
   $$

2. **パラメータ $\mu_{ki}$ に関する偏微分**:
   特定の特徴 $i$ とクラス $k$ に関するパラメータ $\mu_{ki}$ で偏微分します：
   $$
   \frac{\partial \ln L}{\partial \mu_{ki}} = \sum_{n=1}^N t_{nk} \left( \frac{x_{ni}}{\mu_{ki}} - \frac{1 - x_{ni}}{1 - \mu_{ki}} \right) = \sum_{n=1}^N t_{nk} \frac{x_{ni} - \mu_{ki}}{\mu_{ki}(1 - \mu_{ki})}
   $$
   これを 0 とおくと：
   $$
   \sum_{n=1}^N t_{nk} x_{ni} - \mu_{ki} \sum_{n=1}^N t_{nk} = 0 \implies \sum_{n=1}^N t_{nk} x_{ni} - \mu_{ki} N_k = 0
   $$
   したがって：
   $$
   \mu_{ki} = \frac{1}{N_k} \sum_{n=1}^N t_{nk} x_{ni}
   $$
   が得られます。これはクラス $\mathcal{C}_k$ に属するサンプルのうち特徴 $i$ が 1 であるサンプルの比率（標本平均）に正確に一致します。 $\blacksquare$


In [16]:
# Exercise 5.15 数値検証
X_bin = np.array([
    [1, 0, 1],
    [1, 1, 1],
    [0, 1, 0],
    [0, 0, 1]
])
t_bin = np.array([0, 0, 1, 1])

mu_est = exercise_5_15_mle_binary_naive_bayes(X_bin, t_bin, 2)
print("クラス0 特徴生起確率 mu_0i:", mu_est[0])
print("クラス1 特徴生起確率 mu_1i:", mu_est[1])

assert np.allclose(mu_est[0], [1.0, 0.5, 1.0])
assert np.allclose(mu_est[1], [0.0, 0.5, 0.5])
print("Exercise 5.15: 検証成功！")


クラス0 特徴生起確率 mu_0i: [1.  0.5 1. ]
クラス1 特徴生起確率 mu_1i: [0.  0.5 0.5]
Exercise 5.15: 検証成功！


---
### Exercise 5.16: マルチステートナイーブベイズの事前活性化の線形性
**Problem 5.16 (? ?)**:
> Consider a classification problem with $K$ classes for which the feature vector $\boldsymbol{\phi}$ has $M$ components each of which can take $L$ discrete states. Let the values of the components be represented by a 1-of-$L$ binary coding scheme. Further suppose that, conditioned on the class $\mathcal{C}_k$, the $M$ components of $\boldsymbol{\phi}$ are independent. Show that the quantities $a_k$, which appear in the argument to the softmax function, are linear functions of the components of $\boldsymbol{\phi}$.

#### 数式導出と論理ステップ:
1. **マルチステート 1-of-$L$ 符号化の定式化**:
   各成分 $m \in \{1, \dots, M\}$ は $L$ 個の状態をとり、1-of-$L$ 二値表現 $\boldsymbol{\phi}_m = (\phi_{m1}, \dots, \phi_{mL})^T$（$\phi_{ml} \in \{0, 1\}, \sum_{l=1}^L \phi_{ml} = 1$）で表されます。
   条件付き独立性より、クラス条件付き確率はカテゴリカル分布の積になります：
   $$
   p(\boldsymbol{\phi} \mid \mathcal{C}_k) = \prod_{m=1}^M \prod_{l=1}^L \mu_{kml}^{\phi_{ml}}, \quad \text{ただし } \sum_{l=1}^L \mu_{kml} = 1
   $$

2. **事後確率の事前活性化 $a_k$ (式 5.46) の計算**:
   ソフトマックス関数の引数となる量 $a_k$ は：
   $$
   a_k = \ln \left\{ p(\boldsymbol{\phi} \mid \mathcal{C}_k) p(\mathcal{C}_k) \right\} = \ln \pi_k + \ln p(\boldsymbol{\phi} \mid \mathcal{C}_k)
   $$
   対数を展開します：
   $$
   a_k = \ln \pi_k + \sum_{m=1}^M \sum_{l=1}^L \phi_{ml} \ln \mu_{kml}
   $$

3. **入力成分に関する線形性の確認**:
   重み係数を $w_{k, ml} = \ln \mu_{kml}$、バイアスを $w_{k0} = \ln \pi_k$ とおくと：
   $$
   a_k = \sum_{m=1}^M \sum_{l=1}^L w_{k, ml} \phi_{ml} + w_{k0} = \mathbf{w}_k^T \boldsymbol{\phi} + w_{k0}
   $$
   となり、$a_k$ は特徴ベクトル $\boldsymbol{\phi}$ の成分 $\phi_{ml}$ の純粋な線形関数であることが示されました。 $\blacksquare$


In [17]:
# Exercise 5.16 数値検証
M, L = 3, 4
phi_1ofL = np.zeros((M, L))
phi_1ofL[0, 2] = 1.0
phi_1ofL[1, 0] = 1.0
phi_1ofL[2, 3] = 1.0

mu_kml = np.array([
    [0.1, 0.2, 0.5, 0.2],
    [0.6, 0.1, 0.2, 0.1],
    [0.1, 0.1, 0.1, 0.7]
])
pi_k = 0.4

ak_val = exercise_5_16_multistate_naive_bayes_ak(phi_1ofL, mu_kml, pi_k)
expected_ak = np.log(pi_k) + np.log(0.5) + np.log(0.6) + np.log(0.7)

print(f"計算された a_k: {ak_val:.6f}")
print(f"理論的期待値:   {expected_ak:.6f}")

assert np.isclose(ak_val, expected_ak)
print("Exercise 5.16: 検証成功！")


計算された a_k: -2.476938
理論的期待値:   -2.476938
Exercise 5.16: 検証成功！


---
### Exercise 5.17: マルチステートナイーブベイズのラグランジュ最尤解
**Problem 5.17 (? ?)**:
> Derive the maximum likelihood solution for the parameters of the probabilistic naive Bayes classifier described in Exercise 5.16.

#### 数式導出と論理ステップ:
1. **ラグランジュ関数の構築**:
   全訓練データのうちクラス $\mathcal{C}_k$ に属するサンプルの部分集合を考えます。
   パラメータ $\mu_{kml}$ は各 $(k, m)$ について $\sum_{l=1}^L \mu_{kml} = 1$ という規格化制約を満たす必要があります。
   ラグランジュ乗数 $\{\lambda_{km}\}$ を導入した目的関数は：
   $$
   \mathcal{L} = \sum_{n=1}^N \sum_{k=1}^K t_{nk} \sum_{m=1}^M \sum_{l=1}^L \phi_{nml} \ln \mu_{kml} + \sum_{k=1}^K \sum_{m=1}^M \lambda_{km} \left( \sum_{l=1}^L \mu_{kml} - 1 \right)
   $$

2. **停留条件と乗数の決定**:
   $\mu_{kml}$ について偏微分します：
   $$
   \frac{\partial \mathcal{L}}{\partial \mu_{kml}} = \frac{\sum_{n=1}^N t_{nk} \phi_{nml}}{\mu_{kml}} + \lambda_{km} = 0
   $$
   ここで $N_{kml} = \sum_{n=1}^N t_{nk} \phi_{nml}$ はクラス $\mathcal{C}_k$ において特徴 $m$ が状態 $l$ をとったサンプル数です：
   $$
   \mu_{kml} = -\frac{N_{kml}}{\lambda_{km}}
   $$
   制約式 $\sum_{l=1}^L \mu_{kml} = 1$ に代入すると：
   $$
   \sum_{l=1}^L \left(-\frac{N_{kml}}{\lambda_{km}}\right) = -\frac{1}{\lambda_{km}} \sum_{l=1}^L N_{kml} = -\frac{N_k}{\lambda_{km}} = 1 \implies \lambda_{km} = -N_k
   $$

3. **最尤推定量の導出**:
   $\lambda_{km} = -N_k$ を代入することで：
   $$
   \mu_{kml} = \frac{N_{kml}}{N_k}
   $$
   が得られます。これは直感通り、クラス $\mathcal{C}_k$ において特徴 $m$ が状態 $l$ をとった頻度比率に厳密に一致します。 $\blacksquare$


In [18]:
# Exercise 5.17 数値検証
# クラスkのデータ N_k = 5, 特徴数 M = 2, 状態数 L = 3
features_sample = np.zeros((5, 2, 3))
states = [[0, 2], [0, 1], [1, 2], [0, 2], [2, 0]]
for n, (s0, s1) in enumerate(states):
    features_sample[n, 0, s0] = 1.0
    features_sample[n, 1, s1] = 1.0

mle_mu = exercise_5_17_mle_multistate_naive_bayes(features_sample)
print("特徴0 の状態確率 mu_k0l:", mle_mu[0])
print("特徴1 の状態確率 mu_k1l:", mle_mu[1])

assert np.allclose(mle_mu[0], [3/5, 1/5, 1/5])
assert np.allclose(mle_mu[1], [1/5, 1/5, 3/5])
assert np.isclose(np.sum(mle_mu[0]), 1.0)
assert np.isclose(np.sum(mle_mu[1]), 1.0)
print("Exercise 5.17: 検証成功！")


特徴0 の状態確率 mu_k0l: [0.6 0.2 0.2]
特徴1 の状態確率 mu_k1l: [0.2 0.2 0.6]
Exercise 5.17: 検証成功！


## Part 4: 5.4節 識別モデル (Discriminative Classifiers)


---
### Exercise 5.18: ロジスティック・シグモイド関数の導関数の検証
**Problem 5.18 (?)**:
> Verify the relation (5.72) for the derivative of the logistic sigmoid function defined by (5.42).

#### 数式導出と論理ステップ:
1. **微分の計算**:
   $\sigma(a) = (1 + e^{-a})^{-1}$ に対する商の微分公式または合成関数の微分公式より：
   $$
   \frac{d\sigma}{da} = -(1 + e^{-a})^{-2} \cdot \frac{d}{da}(1 + e^{-a}) = -(1 + e^{-a})^{-2} (-e^{-a}) = \frac{e^{-a}}{(1 + e^{-a})^2}
   $$

2. **積の形への変形**:
   右辺を分解します：
   $$
   \frac{e^{-a}}{(1 + e^{-a})^2} = \frac{1}{1 + e^{-a}} \cdot \frac{e^{-a}}{1 + e^{-a}} = \sigma(a) \cdot \frac{(1 + e^{-a}) - 1}{1 + e^{-a}} = \sigma(a) \left( 1 - \frac{1}{1 + e^{-a}} \right)
   $$
   したがって：
   $$
   \frac{d\sigma}{da} = \sigma(a) (1 - \sigma(a)) \tag{5.72}
   $$
   が示されました。 $\blacksquare$


In [19]:
# Exercise 5.18 数値検証
a_pts = np.array([-3.0, -1.0, 0.0, 1.0, 3.0])
deriv_verified = exercise_5_18_verify_sigmoid_derivative(a_pts)

eps = 1e-7
num_deriv = (sigmoid(a_pts + eps) - sigmoid(a_pts - eps)) / (2 * eps)
ana_deriv = sigmoid_deriv(a_pts)

print("解析的導関数:", ana_deriv)
print("数値微分値:  ", num_deriv)

assert deriv_verified
assert np.allclose(ana_deriv, num_deriv, atol=1e-6)
print("Exercise 5.18: 検証成功！")


解析的導関数: [0.04517666 0.19661193 0.25       0.19661193 0.04517666]
数値微分値:   [0.04517666 0.19661193 0.25       0.19661193 0.04517666]
Exercise 5.18: 検証成功！


---
### Exercise 5.19: ロジスティック回帰の交差エントロピー誤差関数の勾配導出
**Problem 5.19 (?)**:
> By making use of the result (5.72) for the derivative of the logistic sigmoid, show that the derivative of the error function (5.74) for the logistic regression model is given by (5.75).

#### 数式導出と論理ステップ:
1. **交差エントロピー誤差関数 (式 5.74)**:
   $$
   E(\mathbf{w}) = -\sum_{n=1}^N \left\{ t_n \ln y_n + (1 - t_n) \ln(1 - y_n) \right\}, \quad y_n = \sigma(a_n), \; a_n = \mathbf{w}^T \boldsymbol{\phi}_n
   $$

2. **連鎖律の適用**:
   合成関数の微分法により：
   $$
   \nabla E(\mathbf{w}) = \sum_{n=1}^N \frac{\partial E_n}{\partial y_n} \frac{d y_n}{d a_n} \nabla_{\mathbf{w}} a_n
   $$
   各因子を個別に計算します：
   - $\frac{\partial E_n}{\partial y_n} = -\frac{t_n}{y_n} + \frac{1 - t_n}{1 - y_n} = \frac{y_n - t_n}{y_n(1 - y_n)}$
   - $\frac{d y_n}{d a_n} = \frac{d\sigma}{d a_n} = y_n(1 - y_n)$ (式 5.72 より)
   - $\nabla_{\mathbf{w}} a_n = \boldsymbol{\phi}_n$

3. **シグモイド微分の相殺**:
   積をとると分母の $y_n(1 - y_n)$ が驚くほど見事に相殺されます：
   $$
   \nabla E(\mathbf{w}) = \sum_{n=1}^N \left( \frac{y_n - t_n}{y_n(1 - y_n)} \right) \left\{ y_n(1 - y_n) \right\} \boldsymbol{\phi}_n = \sum_{n=1}^N (y_n - t_n) \boldsymbol{\phi}_n \tag{5.75}
   $$
   この結果は線形回帰における二乗和誤差の勾配と全く同じ構造を共有しています。 $\blacksquare$


In [20]:
# Exercise 5.19 数値検証
np.random.seed(42)
N, D = 15, 3
Phi_ex19 = np.random.randn(N, D)
t_ex19 = np.random.randint(0, 2, size=N)
w_ex19 = np.random.randn(D)

g_ana, g_num = exercise_5_19_verify_logistic_gradient(Phi_ex19, t_ex19, w_ex19)
print("解析的勾配 sum (y_n - t_n) phi_n:", g_ana)
print("数値微分による中心差分勾配:     ", g_num)

assert np.allclose(g_ana, g_num, atol=1e-5)
print("Exercise 5.19: 検証成功！")


解析的勾配 sum (y_n - t_n) phi_n: [-2.47505539 -2.24579264 -5.98377402]
数値微分による中心差分勾配:      [-2.47505538 -2.24579263 -5.983774  ]
Exercise 5.19: 検証成功！


---
### Exercise 5.20: 線形分離可能データにおける非正則化重みノルムの発散
**Problem 5.20 (?)**:
> Show that for a linearly separable data set, the maximum likelihood solution for the logistic regression model is obtained by finding a vector $\mathbf{w}$ whose decision boundary $\mathbf{w}^T \boldsymbol{\phi}(\mathbf{x}) = 0$ separates the classes and then taking the magnitude of $\mathbf{w}$ to infinity.

#### 数式導出と論理ステップ:
1. **線形分離可能性の数理的表現**:
   データセットが線形分離可能であるとは、ある重みベクトル $\mathbf{w}^*$ が存在して：
   $$
   \forall n, \; \begin{cases} \mathbf{w}^{*T} \boldsymbol{\phi}_n > 0 & (t_n = 1 \text{ のとき}) \\ \mathbf{w}^{*T} \boldsymbol{\phi}_n < 0 & (t_n = 0 \text{ のとき}) \end{cases}
   $$
   が成立することを意味します。統一表記として $(2t_n - 1) \mathbf{w}^{*T} \boldsymbol{\phi}_n > 0$ です。

2. **重みのスケール倍 $\mathbf{w} = k \mathbf{w}^*$ ($k > 0$) の挙動**:
   この方向を保ったまま重みの大きさをスケール係数 $k$ で拡大します。
   - $t_n = 1$ のサンプルでは、$a_n = k \mathbf{w}^{*T} \boldsymbol{\phi}_n \to +\infty$ ($k \to \infty$) となるため：
     $$
     y_n = \sigma(a_n) \to 1 \implies \ln y_n \to 0
     $$
   - $t_n = 0$ のサンプルでは、$a_n = k \mathbf{w}^{*T} \boldsymbol{\phi}_n \to -\infty$ ($k \to \infty$) となるため：
     $$
     y_n = \sigma(a_n) \to 0 \implies \ln(1 - y_n) \to 0
     $$

3. **交差エントロピー誤差関数の極限**:
   全データに対する交差エントロピー誤差は：
   $$
   E(k \mathbf{w}^*) = -\sum_{n=1}^N \left\{ t_n \ln y_n + (1 - t_n) \ln(1 - y_n) \right\} \to 0 \quad (k \to \infty)
   $$
   誤差関数 $E(\mathbf{w}) \ge 0$ は常に正値であり、$E = 0$ は大域的最小値の下限（infimum）です。
   しかし、任意の有限の $\mathbf{w}$ に対して $y_n \in (0, 1)$ であるため $E(\mathbf{w}) > 0$ であり、$E = 0$ に到達するためには $k \to \infty$、すなわち：
   $$
   \|\mathbf{w}\| \to \infty
   $$
   としなければなりません。このため、最尤推定解は発散し、決定境界における事後確率はヘビサイドの階段関数に退化して極端な過適合を引き起こします。 $\blacksquare$


In [21]:
# Exercise 5.20 数値検証
X_sep = np.array([[-2.0, -1.0], [2.0, 1.0]])
t_sep = np.array([0, 1])

iters = [20, 50, 100, 200]
norms = exercise_5_20_separable_growth_check(X_sep, t_sep, iters)

for it, nm in zip(iters, norms):
    print(f"反復回数 {it:3d}: 重みノルム ||w|| = {nm:.4f}")

assert norms[-1] > norms[0]
assert norms[-1] > 2.5
print("Exercise 5.20: 検証成功！")


反復回数  20: 重みノルム ||w|| = 2.1047
反復回数  50: 重みノルム ||w|| = 2.4901
反復回数 100: 重みノルム ||w|| = 2.7907
反復回数 200: 重みノルム ||w|| = 3.0954
Exercise 5.20: 検証成功！


---
### Exercise 5.21: ソフトマックス活性化関数のヤコビ行列の導出
**Problem 5.21 (?)**:
> Show that the derivatives of the softmax activation function (5.76), where the $a_k$ are defined by (5.77), are given by (5.78).

#### 数式導出と論理ステップ:
1. **ソフトマックス関数の定義**:
   $$
   y_k = \frac{e^{a_k}}{\sum_{m=1}^K e^{a_m}} \tag{5.76}
   $$

2. **商の微分法による導関数の計算**:
   活性化 $a_j$ に関して偏微分します：
   $$
   \frac{\partial y_k}{\partial a_j} = \frac{\frac{\partial (e^{a_k})}{\partial a_j} \left(\sum_{m} e^{a_m}\right) - e^{a_k} \frac{\partial}{\partial a_j}\left(\sum_{m} e^{a_m}\right)}{\left( \sum_m e^{a_m} \right)^2}
   $$
   ここで $\frac{\partial (e^{a_k})}{\partial a_j} = I_{kj} e^{a_k}$、および $\frac{\partial}{\partial a_j}\left(\sum_{m} e^{a_m}\right) = e^{a_j}$ です：
   $$
   \frac{\partial y_k}{\partial a_j} = \frac{I_{kj} e^{a_k} \left(\sum_m e^{a_m}\right) - e^{a_k} e^{a_j}}{\left( \sum_m e^{a_m} \right)^2}
   $$

3. **事後確率 $y_k, y_j$ による整理**:
   項ごとに分解します：
   $$
   \frac{\partial y_k}{\partial a_j} = I_{kj} \frac{e^{a_k}}{\sum_m e^{a_m}} - \frac{e^{a_k}}{\sum_m e^{a_m}} \frac{e^{a_j}}{\sum_m e^{a_m}} = I_{kj} y_k - y_k y_j = y_k (I_{kj} - y_j) \tag{5.78}
   $$
   これにより式 (5.78) が厳密に導出されました。 $\blacksquare$


In [22]:
# Exercise 5.21 数値検証
a_vec = np.array([1.2, -0.5, 2.1])
J_ana, J_num = exercise_5_21_verify_softmax_jacobian(a_vec)

print("解析的ヤコビ行列 J_kj = y_k(I_kj - y_j):")
print(J_ana)
print("数値微分ヤコビ行列:")
print(J_num)

assert np.allclose(J_ana, J_num, atol=1e-5)
print("Exercise 5.21: 検証成功！")


解析的ヤコビ行列 J_kj = y_k(I_kj - y_j):
[[ 0.19917356 -0.01377055 -0.18540302]
 [-0.01377055  0.04764062 -0.03387008]
 [-0.18540302 -0.03387008  0.21927309]]
数値微分ヤコビ行列:
[[ 0.19917356 -0.01377055 -0.18540302]
 [-0.01377055  0.04764062 -0.03387008]
 [-0.18540302 -0.03387008  0.21927309]]
Exercise 5.21: 検証成功！


---
### Exercise 5.22: 多クラス交差エントロピー誤差関数の勾配導出
**Problem 5.22 (?)**:
> Using the result (5.78) for the derivatives of the softmax activation function, show that the gradients of the cross-entropy error (5.80) are given by (5.81).

#### 数式導出と論理ステップ:
1. **多クラス交差エントロピー誤差関数 (式 5.80)**:
   $$
   E(\mathbf{W}) = -\sum_{n=1}^N \sum_{k=1}^K t_{nk} \ln y_{nk}, \quad y_{nk} = \frac{e^{a_{nk}}}{\sum_m e^{a_{nm}}}, \; a_{nk} = \mathbf{w}_k^T \boldsymbol{\phi}_n
   $$

2. **活性化 $a_{nj}$ に対する連鎖律**:
   各データ点 $n$ の誤差寄与 $E_n = -\sum_{k=1}^K t_{nk} \ln y_{nk}$ の $a_{nj}$ による偏微分を計算します：
   $$
   \frac{\partial E_n}{\partial a_{nj}} = \sum_{k=1}^K \frac{\partial E_n}{\partial y_{nk}} \frac{\partial y_{nk}}{\partial a_{nj}}
   $$
   $\frac{\partial E_n}{\partial y_{nk}} = -\frac{t_{nk}}{y_{nk}}$、および式 (5.78) より $\frac{\partial y_{nk}}{\partial a_{nj}} = y_{nk}(I_{kj} - y_{nj})$ です：
   $$
   \frac{\partial E_n}{\partial a_{nj}} = \sum_{k=1}^K \left( -\frac{t_{nk}}{y_{nk}} \right) y_{nk} (I_{kj} - y_{nj}) = -\sum_{k=1}^K t_{nk} (I_{kj} - y_{nj})
   $$
   和を展開すると：
   $$
   \frac{\partial E_n}{\partial a_{nj}} = -\sum_{k=1}^K t_{nk} I_{kj} + y_{nj} \sum_{k=1}^K t_{nk} = -t_{nj} + y_{nj} (1) = y_{nj} - t_{nj}
   $$
   （ここで 1-of-$K$ 符号化の性質 $\sum_{k=1}^K t_{nk} = 1$ を用いました）。

3. **重みベクトル $\mathbf{w}_j$ に対する勾配**:
   $a_{nj} = \mathbf{w}_j^T \boldsymbol{\phi}_n$ より $\nabla_{\mathbf{w}_j} a_{nj} = \boldsymbol{\phi}_n$ であるため：
   $$
   \nabla_{\mathbf{w}_j} E(\mathbf{W}) = \sum_{n=1}^N \frac{\partial E_n}{\partial a_{nj}} \boldsymbol{\phi}_n = \sum_{n=1}^N (y_{nj} - t_{nj}) \boldsymbol{\phi}_n \tag{5.81}
   $$
   が導かれました。 $\blacksquare$


In [23]:
# Exercise 5.22 数値検証
np.random.seed(42)
N, D, K = 20, 2, 3
Phi_ex22 = np.random.randn(N, D)
t_raw = np.random.randint(0, K, size=N)
T_ex22 = np.eye(K)[t_raw]
W_ex22 = np.random.randn(D, K)

g_ana, g_num = exercise_5_22_verify_softmax_gradient(Phi_ex22, T_ex22, W_ex22)
print("解析的重み行列勾配 Phi^T (Y - T):")
print(g_ana)
print("数値中心差分勾配:")
print(g_num)

assert np.allclose(g_ana, g_num, atol=1e-4)
print("Exercise 5.22: 検証成功！")


解析的重み行列勾配 Phi^T (Y - T):
[[ 3.67654307 -5.74756266  2.07101959]
 [-0.72405835 -3.14396195  3.86802029]]
数値中心差分勾配:
[[ 3.67654307 -5.74756266  2.07101959]
 [-0.72405835 -3.14396194  3.8680203 ]]
Exercise 5.22: 検証成功！


---
### Exercise 5.23: プロビット関数と誤差関数 (erf) の等価性
**Problem 5.23 (?)**:
> Show that the probit function (5.86) and the erf function (5.87) are related by (5.88).

#### 数式導出と論理ステップ:
1. **プロビット関数と誤差関数の定義**:
   標準正規累積分布関数 (式 5.86):
   $$
   \Phi(a) = \int_{-\infty}^a \mathcal{N}(\theta \mid 0, 1) d\theta = \frac{1}{\sqrt{2\pi}} \int_{-\infty}^a \exp\left( -\frac{\theta^2}{2} \right) d\theta
   $$
   誤差関数 (式 5.87):
   $$
   \operatorname{erf}(x) = \frac{2}{\sqrt{\pi}} \int_0^x \exp(-u^2) du
   $$

2. **積分の分割と変数変換**:
   ガウス関数の対称性 $\int_{-\infty}^0 \mathcal{N}(\theta \mid 0, 1) d\theta = \frac{1}{2}$ を用いて積分区間を分割します：
   $$
   \Phi(a) = \int_{-\infty}^0 \mathcal{N}(\theta \mid 0, 1) d\theta + \int_0^a \mathcal{N}(\theta \mid 0, 1) d\theta = \frac{1}{2} + \frac{1}{\sqrt{2\pi}} \int_0^a \exp\left(-\frac{\theta^2}{2}\right) d\theta
   $$
   変数変換 $u = \frac{\theta}{\sqrt{2}}$（したがって $\theta = \sqrt{2}u, \; d\theta = \sqrt{2} du$）を行うと：
   $$
   \frac{1}{\sqrt{2\pi}} \int_0^a \exp\left(-\frac{\theta^2}{2}\right) d\theta = \frac{1}{\sqrt{2\pi}} \int_0^{a/\sqrt{2}} \exp(-u^2) (\sqrt{2} du) = \frac{1}{\sqrt{\pi}} \int_0^{a/\sqrt{2}} \exp(-u^2) du
   $$

3. **誤差関数の代入**:
   誤差関数の定義式と照合すると：
   $$
   \frac{1}{\sqrt{\pi}} \int_0^{a/\sqrt{2}} \exp(-u^2) du = \frac{1}{2} \left( \frac{2}{\sqrt{\pi}} \int_0^{a/\sqrt{2}} \exp(-u^2) du \right) = \frac{1}{2} \operatorname{erf}\left( \frac{a}{\sqrt{2}} \right)
   $$
   したがって：
   $$
   \Phi(a) = \frac{1}{2} \left\{ 1 + \operatorname{erf}\left( \frac{a}{\sqrt{2}} \right) \right\} \tag{5.88}
   $$
   が完全に導出されました。 $\blacksquare$


In [24]:
# Exercise 5.23 数値検証
a_vals = np.linspace(-3.0, 3.0, 30)
verified_erf = exercise_5_23_verify_probit_erf(a_vals)

probit_vals = probit(a_vals)
erf_formula_vals = 0.5 * (1.0 + special.erf(a_vals / np.sqrt(2.0)))

assert verified_erf
assert np.allclose(probit_vals, erf_formula_vals)
print("Exercise 5.23: 検証成功！")


Exercise 5.23: 検証成功！


---
### Exercise 5.24: プロビットとシグモイドの原点傾き一致条件
**Problem 5.24 (? ?)**:
> Suppose we wish to approximate the logistic sigmoid $\sigma(a)$ defined by (5.42) by a scaled probit function $\Phi(\lambda a)$, where $\Phi(a)$ is defined by (5.86). Show that if $\lambda$ is chosen so that the derivatives of the two functions are equal at $a = 0$, then $\lambda^2 = \pi/8$.

#### 数式導出と論理ステップ:
1. **原点におけるロジスティック・シグモイドの微分**:
   シグモイドの導関数は $\sigma'(a) = \sigma(a)(1 - \sigma(a))$ です。
   原点 $a = 0$ において $\sigma(0) = \frac{1}{1 + e^0} = \frac{1}{2}$ であるため：
   $$
   \sigma'(0) = \sigma(0)(1 - \sigma(0)) = \frac{1}{2}\left(1 - \frac{1}{2}\right) = \frac{1}{4}
   $$

2. **原点におけるスケールドプロビットの微分**:
   関数 $g(a) = \Phi(\lambda a)$ の微分を合成関数の微分法により計算します：
   $$
   g'(a) = \frac{d}{da} \Phi(\lambda a) = \lambda \Phi'(\lambda a)
   $$
   $\Phi(u)$ は標準正規分布の累積分布関数であるため、その導関数は標準正規確率密度関数です：
   $$
   \Phi'(u) = \mathcal{N}(u \mid 0, 1) = \frac{1}{\sqrt{2\pi}} \exp\left( -\frac{u^2}{2} \right)
   $$
   したがって：
   $$
   g'(a) = \frac{\lambda}{\sqrt{2\pi}} \exp\left( -\frac{\lambda^2 a^2}{2} \right)
   $$
   原点 $a = 0$ を代入すると：
   $$
   g'(0) = \frac{\lambda}{\sqrt{2\pi}} \exp(0) = \frac{\lambda}{\sqrt{2\pi}}
   $$

3. **原点傾き一致方程式の求解**:
   両者の原点における微分値を等しいとおきます：
   $$
   \sigma'(0) = g'(0) \implies \frac{1}{4} = \frac{\lambda}{\sqrt{2\pi}}
   $$
   $\lambda$ について解くと：
   $$
   \lambda = \frac{\sqrt{2\pi}}{4}
   $$
   両辺を2乗すると：
   $$
   \lambda^2 = \frac{2\pi}{16} = \frac{\pi}{8}
   $$
   数値的には $\lambda = \sqrt{\pi / 8} \approx 0.626657$ です。
   これにより、スケールドプロビット関数 $\Phi(\sqrt{\pi/8} \, a)$ がシグモイド関数 $\sigma(a)$ の極めて高精度な解析的近似を与える理由が厳密に証明されました。 $\blacksquare$


In [25]:
# Exercise 5.24 数値検証
ds0, dp0, lambda_sq = exercise_5_24_probit_sigmoid_matching_scale()

print(f"原点における sigma'(0):                 {ds0:.6f}")
print(f"原点における d/da Phi(lambda*a)|_{{a=0}}: {dp0:.6f}")
print(f"導出された lambda^2:                   {lambda_sq:.6f}")
print(f"理論値 pi / 8:                         {np.pi / 8.0:.6f}")

assert np.isclose(ds0, dp0)
assert np.isclose(lambda_sq, np.pi / 8.0)

# 可視化検証: シグモイドとスケールドプロビットの比較プロット
a_grid = np.linspace(-4, 4, 100)
fig, ax = plt.subplots(figsize=(7, 4.5))
ax.plot(a_grid, sigmoid(a_grid), label=r"$\sigma(a)$ (Logistic Sigmoid)", lw=2.5, color='crimson')
ax.plot(a_grid, probit(np.sqrt(np.pi / 8.0) * a_grid), '--', label=r"$\Phi(\sqrt{\pi/8} a)$ (Scaled Probit)", lw=2.0, color='dodgerblue')
ax.set_title(r"Comparison of Logistic Sigmoid and Scaled Probit ($\lambda^2 = \pi/8$)", fontsize=13)
ax.set_xlabel("a", fontsize=11)
ax.set_ylabel("Activation", fontsize=11)
ax.grid(True, alpha=0.3)
ax.legend(fontsize=11)
fig.tight_layout()
plt.close(fig)

print("Exercise 5.24: 検証成功！")


原点における sigma'(0):                 0.250000
原点における d/da Phi(lambda*a)|_{a=0}: 0.250000
導出された lambda^2:                   0.392699
理論値 pi / 8:                         0.392699


Exercise 5.24: 検証成功！


---
## 第5章 演習問題 総括 (Chapter 5 Exercises Summary)

本ノートブックでは、『深層学習：基礎と概念 (Bishop & Bishop 2024)』第5章の**全24問 (Exercises 5.1 〜 5.24)** について、以下の4大テーマにわたる厳密な数学的証明と数値検証を完遂しました：

1. **線形判別モデル (5.1節, Ex 5.1 - 5.4)**:
   - 1-of-$K$ 二値符号化における条件付き期待値 $\mathbb{E}[\mathbf{t} \mid \mathbf{x}]$ がベイズ事後確率 $p(\mathcal{C}_k \mid \mathbf{x})$ に一致することの証明。
   - 凸包 (Convex Hull) の交差と線形分離可能性の同値性（強分離超平面定理）。
   - バイアス項 $\phi_0(\mathbf{x})=1$ の存在下で、訓練目標値が満たす単一または複数の線形制約が、最小二乗法の未知予測値において恒等的に保存されることの代数的証明。
2. **決定理論 (5.2節, Ex 5.5 - 5.10)**:
   - 適合率と再現率の重み付き調和平均からの $F_\beta$ スコア陽公式の代数導出。
   - 非負数の相乗平均不等式を用いたバタチャリア誤分類確率上界 $p(\text{mistake}) \le \int \{p(\mathbf{x}, \mathcal{C}_1)p(\mathbf{x}, \mathcal{C}_2)\}^{1/2} d\mathbf{x}$ の導出。
   - 対称 0-1 損失行列における期待リスク最小化が MAP (事後確率最大化) 決定規則と厳密に等価であることの証明。
   - 一般損失行列および事前確率の下での期待損失最小化基準の導出。
   - サンプル事後確率の標本平均の大数の強法則によるクラス事前確率 $p(\mathcal{C}_k)$ への漸近収束証明。
   - 棄却損失 $\lambda$ を含む棄却オプション付き決定基準の導出と棄却閾値との関係式 $\theta = 1 - \lambda$ の確立。
3. **生成モデル (5.3節, Ex 5.11 - 5.17)**:
   - ロジスティック・シグモイド関数の点対称性 $\sigma(-a) = 1 - \sigma(a)$ および逆関数ロジットの導出。
   - 2クラス共通共分散ガウス生成モデルの事後確率ロジスティック形式 $a = \mathbf{w}^T \mathbf{x} + w_0$ の二次項相殺とパラメータ解析解の導出。
   - ラグランジュ未定乗数法による多クラス事前確率最尤推定量 $\pi_k = N_k / N$ の導出。
   - 共通共分散ガウス生成モデルにおけるクラス平均 $\boldsymbol{\mu}_k$ および加重平均共通共分散行列 $\boldsymbol{\Sigma} = \sum (N_k/N)\mathbf{S}_k$ の行列微分による最尤推定。
   - 離散2値ベルヌーイナイーブベイズの最尤解 $\mu_{ki} = \frac{1}{N_k}\sum t_{nk}x_{ni}$ の導出。
   - $M$ 成分 $L$ 状態離散特徴をもつマルチステートナイーブベイズの事前活性化 $a_k$ が 1-of-$L$ 特徴量の線形関数になることの証明と、そのパラメータ最尤推定解 $\mu_{kml} = N_{kml}/N_k$ の導出。
4. **識別モデル (5.4節, Ex 5.18 - 5.24)**:
   - ロジスティック・シグモイド関数の導関数公式 $\frac{d\sigma}{da} = \sigma(1 - \sigma)$ の証明。
   - ロジスティック回帰の交差エントロピー誤差関数の勾配 $\nabla E = \sum_n (y_n - t_n)\boldsymbol{\phi}_n$ におけるシグモイド微分の分母相殺の証明。
   - 線形分離可能データにおける非正則化ロジスティック回帰の最尤解の重みノルム発散 $\|\mathbf{w}\| \to \infty$ と過適合・過信の発生機構の証明。
   - ソフトマックス活性化関数のヤコビ行列 $\frac{\partial y_k}{\partial a_j} = y_k(I_{kj} - y_j)$ のクロネッカーのデルタを用いた統一導出。
   - 多クラス交差エントロピー誤差関数の重み行列勾配 $\nabla_{\mathbf{w}_j} E = \sum_n (y_{nj} - t_{nj})\boldsymbol{\phi}_n$ の導出。
   - プロビット関数 $\Phi(a)$ と誤差関数 $\operatorname{erf}(x)$ の恒等関係 $\Phi(a) = \frac{1}{2}[1 + \operatorname{erf}(a/\sqrt{2})]$ の変数変換による証明。
   - ロジスティック・シグモイド $\sigma(a)$ とスケールドプロビット $\Phi(\lambda a)$ の原点での微分一致条件 $\lambda^2 = \pi / 8$ の導出。

全24問の自己検証テストセルおよびアサーションがすべて成功裏に実行され、理論的厳密性と数値的無矛盾性が完全に実証されました。
